---
title: "Bioxrvir_with_ANANSE_24"
author: "Klara Dewitte"
date: "2022-11-28"
output: html_document
---
##Load packages, get RNA-seq data and get DE genes
Load the packages.
```{r libraries}
rm(list = ls())
library(limma) 
library(edgeR)
library(stringr)
library(tidyverse)
library(dplyr)
library(nichenetr)
library(biomaRt)
library(data.table)
library(stringr)
library(synapser) 
library(knitr)
library(ggstatsplot)
library(cowplot)
library(patchwork)
memory.size(1000000)
```

Get the RNAsesq data from Synapsr.
```{r data, echo=TRUE}
# login of synapsr
synLogin('KlaraDewitte','Thesis2022') 
 
# Obtain a pointer and download the RNA-seq level 1 data
syn18508389 <- synGet(entity = 'syn18508389') 
filepath_RNA <- syn18508389$path
rawdata <- read.csv(filepath_RNA, row.names = 1)

# Obtain pointer and download metadata
ssyn15574180 <- synGet(entity = 'syn15574180') 
filepath_metadata <- ssyn15574180$path
metadata <- read.csv(filepath_metadata)

# Obtain the different conditions at 24h
values <- stringr::str_split(metadata$specimenName, "_") %>% 
                  do.call(rbind, .) 

# Obtain all the ids of samples at 24h
id_24 <- metadata[which(values[,2] == 24 & metadata$RNAseq_QCpass == TRUE) ,]$specimenID 

# Subset the raw RNAseq data and metadata for 24h
rawdata_subset_24 <- rawdata[,which(colnames(rawdata) %in% id_24)]
metadata_subset_24 <- metadata[which(metadata$specimenID %in% id_24),]

# Get unique values for each cytokine, timepoint and replicate
unique_values <- values %>%
                  apply(2, unique) %>%
                  setNames(c("cytokine", "timepoint", "collection", "replicate"))

# Connect to Ensembl database
ensembl <- useEnsembl(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")

# Get treatment groups for 24h
Treat <- sapply(str_split(metadata_subset_24$specimenName, "_"), function(k) k[1])
Treat <- Treat %>% factor %>% relevel(ref = "PBS")

# Create design matrix based on treatment
design <- model.matrix(~ Treat)
rownames(design) <- metadata_subset_24$specimenName
    
# Preprocess counts, estimate dispersion
y <- DGEList(counts = rawdata_subset_24, group = Treat)
y <- calcNormFactors(y)
y <- estimateDisp(y, design, robust = TRUE)
    
# Fit the model
fit <- glmQLFit(y, design)
    
# Perform test for each cytokine against PBS
res <- data.table()
for (treat in colnames(design)[-1]){
  qlf <- glmQLFTest(fit, coef = treat)
  DE_genes_table <- data.frame(topTags(qlf, sort.by = "PValue", n = Inf),  coef = substring(treat, 6))
  DE_genes_table$Gene <- rownames(DE_genes_table)
  res <- rbind(res, DE_genes_table)
}
   
# Add gene names
IDtoSymbol_matrix <- getBM(attributes = c("ensembl_gene_id", "external_gene_name"), filters = "ensembl_gene_id", values = unique(res$Gene), mart = ensembl)
top_table <- merge(res, IDtoSymbol_matrix, by.x = 'Gene', by.y = 'ensembl_gene_id', all.x = TRUE)
    
# Omit all the genes with NA or an empty name
top_na <- na.omit(top_table) 
top_na <- top_na[!(top_na$external_gene_name==""),]
    
# Select DE genes and make expression_settings_validation for evaluation NicheNet
total <- c()
total_names_DE <- c()
DEgenes <- list
expression_settings_validation_DE <- list()
DE_info <- list()
designs <- list()
table1 <- list()

for(treatx in unique_values$cytokine){
  if (treatx != "PBS" & treatx != "ctrl"){
    top_na_subset <- top_na[top_na$coef == treatx,]
    DE <- which(top_na_subset$FDR < 0.05 & abs(top_na_subset$logFC) > 2)
    DEgenes <- top_na_subset[DE,]
        
    table2 <- list()
    table2$DEgenes <- DEgenes
    DE_info <- append(DE_info, list(DEgenes))
    
    treatx <- gsub("TGFB", "TGFB1", treatx)
    treatx <- gsub("IFNg", "IFNG", treatx)
        
    print(treatx)
    print(paste0("number of DE genes ", length(DE)))
    
    if (length(DE) > 0){
      total_names_DE <- append(total_names_DE, treatx)
      subsets <- subset(top_na_subset, select = c(logFC, FDR, external_gene_name))
      names(subsets) <- c('lfc', 'qval', 'gene')
      rownames(subsets) <- 1:nrow(subsets)
      table3 <- list()
      table3$diffexp <- as.data.frame(subsets)
      table3$from <- c(treatx)
      table3$name <- c(treatx)
          
      expression_settings_validation_DE <- append(expression_settings_validation_DE, list(table3))
    }
    total <- (append(total, treatx))
    names(expression_settings_validation_DE) <- total_names_DE
    table1$expression_settings_validation_DE <- expression_settings_validation_DE 
        
   }
}
names(DE_info) <- total
table1$DE_info <- DE_info
designs <- append(designs, list(table1))
```

# Evaluation without ATAC-seq
## NicheNet without any modifications
### Ligand_target_matrix included in NicheNet
```{r loading ligand_target_matrices, echo=TRUE}
ligand_target_matrix <- readRDS(url("https://zenodo.org/record/7074291/files/ligand_target_matrix_nsga2r_final.rds"))
```

### Transcriptional response prediction evaluation without ATAC-seq, with NicheNet ligand_target_matrix

```{r model evaluation without ATAC-seq NicheNet matrix, echo=TRUE, error=TRUE, warning=TRUE}
# Extract the gene expression dataset for EGF, HGF and OSM
expression_settings_validation_DE_one <- list(designs[[1]]$expression_settings_validation_DE$EGF,
                                designs[[1]]$expression_settings_validation_DE$HGF,
                                designs[[1]]$expression_settings_validation_DE$OSM)

# Accordingly name the gene expression dataset
names_one <- c("EGF", "HGF", "OSM")
names(expression_settings_validation_DE_one) <- names_one

if(length(expression_settings_validation_DE_one) > 0){
  # Step 1: convert expression datasets to the required format to perform target gene prediction
  settings_one <- lapply(expression_settings_validation_DE_one, function(x) {
  tryCatch({
    convert_expression_settings_evaluation(x)
  }, error = function(e) {
    print(paste("Error occurred in:", x))  # Print the problematic element
    message("Error message:", e)  # Print the error message
    NULL  # Return NULL or any value indicating an issue
  })
})
  
  settings_one = expression_settings_validation_DE_one %>% lapply(convert_expression_settings_evaluation)
  settings_one = settings_one %>% discard(~length(.$from) > 1)
  designs[[1]]$settings_one <- settings_one
  
  # Step 2: calculate the target gene prediction performances
  performances_one = settings_one %>% lapply(evaluate_target_prediction, ligand_target_matrix) %>% bind_rows() 
  designs[[1]]$performances_one <- as.data.frame(performances_one)
    
  # Visualize some classification evaluation metrics showing the target gene prediction performance
  performances_NicheNet = performances_one %>% dplyr::select(-aupr, -auc_iregulon, -pearson_log_pval, -spearman_log_pval,
                                                             -sensitivity_roc, -specificity_roc) %>% 
    gather(key = scorename, value = scorevalue, auroc:spearman)
  
  # Add labels to your measures
  scorelabels = c(auroc = "AUROC", aupr_corrected = "AUPR (corrected)", auc_iregulon_corrected = "AUC-iRegulon (corrected)", 
                  pearson = "Pearson correlation", spearman = "Spearman's rank correlation", 
                  mean_rank_GST_log_pval = "Mean-rank gene-set enrichment")
  
  # Add a random score, to see if the dataset performs better than a random dataset
  scorerandom = c(auroc = 0.5, aupr_corrected = 0, auc_iregulon_corrected = 0, pearson = 0, spearman = 0, 
                  mean_rank_GST_log_pval = 0) %>% data.frame(scorevalue = .) %>% rownames_to_column("scorename")
    
  # Visualistation in the vignette
  print(performances_NicheNet %>%
          mutate(model = "NicheNet v2") %>%
          ggplot() +
          geom_violin(aes(model, scorevalue, group = model, fill = model)) +
          geom_boxplot(aes(model, scorevalue, group = model), width = 0.05) +
          scale_y_continuous("Score target prediction") +
          facet_wrap(~scorename, scales = "free", labeller = as_labeller(scorelabels)) +
          geom_hline(aes(yintercept = scorevalue), data = scorerandom, linetype = 2, color = "red") +
          theme_bw()
  )
  
  # Visualize some classification evaluation metrics showing the target gene prediction performance
  # Choose the measures to include in your visualisation
  performances_NicheNet = performances_one %>% dplyr::select(-aupr, -auc_iregulon, -pearson_log_pval, -spearman_log_pval,
                                                             -sensitivity_roc, -specificity_roc, -mean_rank_GST_log_pval, -auc_iregulon_corrected) %>% 
    gather(key = scorename, value = scorevalue, auroc:spearman)
  
  # Add nice labels to your measures
  scorelabels = c(auroc = "AUROC", aupr_corrected = "AUPR (corrected)",
                  pearson = "Pearson correlation", spearman = "Spearman's rank correlation")
  
  # Add a random score, to see if the dataset performs better than a random dataset
  scorerandom = c(auroc = 0.5, aupr_corrected = 0, pearson = 0, spearman = 0) %>% data.frame(scorevalue = .) %>% rownames_to_column("scorename")
  
  # Manually adapt the colors
  my_colors <- c("red", "blue", "green")

  # Visualize in boxplots, dots per ligand
  p_n <- performances_NicheNet %>%
  mutate(model = "NicheNet v2") %>%
  ggplot() +
  geom_boxplot(aes(model, scorevalue, group = model), width = 0.5, outlier.shape = NA) +
  geom_point(aes(model, scorevalue, color = setting), position = position_jitter(width = 0.2), alpha = 0.7) +
  scale_color_manual(values = my_colors) + 
  scale_y_continuous("Score target prediction") +
  facet_wrap(~scorename, scales = "free", labeller = labeller(scorename = scorelabels)) + 
  geom_hline(aes(yintercept = scorevalue), data = scorerandom, linetype = 2, color = "red") +
  theme_bw() +
  labs(title = "Transcriptional response prediction evaluation 24h", color = "", x = "") +  
  theme(plot.title = element_text(size = 15, hjust = 0.5, face = "bold"), strip.text = element_text(face = "bold"))
  
  #Save the plot for later use
  ggsave("p_n_24.png", plot = p_n, width = 7, height = 7)
}
```

### ligand activity prediction evaluation without ATAC-seq

```{r ligand activity prediction evaluation without ATAC-seq NicheNet, echo=TRUE, error=TRUE}
# Extract the correct expression dataset
designs[[1]]$settings_one <- settings_one
names(settings_one) <- names_one
    
if(length(settings_one) > 0){
  # convert expression datasets to correct format for ligand activity prediction
  all_ligands <- settings_one %>% extract_ligands_from_settings(combination = FALSE) %>% unlist()
  settings_ligand_prediction_one <- settings_one %>% convert_settings_ligand_prediction(all_ligands = all_ligands, validation = TRUE)
    
   # infer ligand importances: for all ligands of interest, we assess how well a ligand explains the differential expression in a specific dataset (and we do this for all datasets).
   ligand_importances_one <- settings_ligand_prediction_one %>% lapply(get_single_ligand_importances, ligand_target_matrix) %>% bind_rows()
   
   ligand_importances_one <- do.call(data.frame, lapply(ligand_importances_one, function(value) replace(value, is.infinite(value), 100000)))
    
    # Look at predictive performance of single/individual importance measures to predict ligand activity: of all ligands tested, the ligand that is truly active in a dataset should get the highest activity score (i.e. best target gene prediction performance)
   evaluation_ligand_prediction_NicheNet <- ligand_importances_one$setting %>% unique() %>% lapply(function(x){x}) %>% 
    lapply(wrapper_evaluate_single_importances_ligand_prediction, ligand_importances_one) %>%
    bind_rows() %>% inner_join(ligand_importances_one %>% distinct(setting, ligand))
   
      # Visualize some classification evaluation metrics showing the ligand activity prediction performance
  evaluation_ligand_prediction_NicheNet <- evaluation_ligand_prediction_NicheNet %>% dplyr::select(-aupr, -sensitivity_roc, -specificity_roc,
                                                                                        -pearson, -spearman,
                                                                                        -mean_rank_GST_log_pval) %>% 
    gather(key = scorename, value = scorevalue, auroc:aupr_corrected)
  
  #Visualisation from Vignette
  scorelabels = c(auroc = "AUROC", aupr_corrected = "AUPR (corrected)")
  
  # Add random score
  scorerandom = c(auroc = 0.5, aupr_corrected = 0) %>% data.frame(scorevalue = .) %>% rownames_to_column("scorename")
    
  print(evaluation_ligand_prediction_NicheNet %>%
     filter(importance_measure %in% c("auroc", "aupr_corrected", "pearson",
                                      "spearman")) %>%
    ggplot() +
    geom_violin(aes(importance_measure, scorevalue, group = importance_measure, fill = importance_measure)) +
    geom_boxplot(aes(importance_measure, scorevalue, group = importance_measure), width = 0.1) +
    scale_y_continuous("Evaluation ligand activity prediction") +
    scale_x_discrete("Ligand activity measure") +
    facet_wrap(~scorename, scales = "free", labeller = as_labeller(scorelabels)) +
    geom_hline(aes(yintercept = scorevalue), data = scorerandom, linetype = 2, color = "red") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90)) +
    ggtitle("NicheNet 24h"))
  
  # Visualisation thesis
  # Select metrics to visualize
  metrics <- c("auroc", "aupr_corrected", "pearson", "spearman")
  my_colors <- c("red", "blue", "green")

  # Create a list of plots for specific metrics
  plots <- lapply(metrics, function(metric) {
  ggplot(ligand_importances_one, aes(x = setting, y = .data[[metric]])) +
    geom_boxplot() +
    geom_jitter(aes(color = factor(test_ligand)), width = 0.2, alpha = 0.7,  size = 3) +
    labs(title = metric,
         x = " ",
         y = metric) +
    scale_color_manual(values = my_colors) + 
    theme_minimal() +
    theme(plot.title = element_text(size = 20, hjust = 0.5),
          plot.margin = margin(1, 1, 1, 1, "cm"),
          legend.position = "none",
      axis.title = element_text(size = 18),  
      axis.text = element_text(size = 18),   
      axis.ticks = element_line(size = 1),  #
      strip.text = element_text(size = 18),
      axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
    theme(legend.title = element_blank(), legend.text = element_text(size = 20), axis.text.x = element_text(angle = 0, hjust = 0.5))
    })
  
  # Combine plots 
  combined_plots <- wrap_plots(plots, ncol = 2) +
  plot_annotation(title = "Ligand activity prediction evaluation 24h", theme = theme(plot.title = element_text(size = 24, hjust = 0.5))) +
  theme(legend.position = "bottom") 
  
  combined_plots
  ggsave("ligand_activity_NicheNet_24.png", plot = combined_plots, width = 13, height = 12, dpi = 300)

}

```


##Construction of ligand_target_matrix
In GRN.R are the weighted networks and individual ligand-target-matrices constructed.

### Construct ligand_target_matrix ATAC+RNA
```{r ligand-target-matrix construction 2, echo=TRUE}
# NicheNet default networks
lr_network <- readRDS(url("https://zenodo.org/record/7074291/files/lr_network_human_21122021.rds"))
sig_network <- readRDS(url("https://zenodo.org/record/7074291/files/signaling_network_human_21122021.rds"))
gr_network <- readRDS(url("https://zenodo.org/record/7074291/files/gr_network_human_21122021.rds"))

##HGF
GRN_HGF_24_ATAC_RNA <- read.delim2("C:/Users/klara/MaStat/Thesis/GithubThesis/Thesis_NicheNet/Ananse_networks/GRN/LINCS_MCF10A_HGF_24_grn.txt", header = TRUE, sep = ";")

# adding our GRN as a new source to the optimized weights
new_gr_network <- gr_network %>% bind_rows(GRN_HGF_24_ATAC_RNA)

new_network_weights_df <- tibble(
  source = "LINCS_MCF10A_HGF_24", 
  avg_weight = 1, 
  median_weight = 1)

new_source_weights_df <- 
  optimized_source_weights_df %>%
  bind_rows(new_network_weights_df)

# aggregate the individual data sources in a weighted manner to obtain a weighted integrated signaling network
weighted_networks <- construct_weighted_networks(
  lr_network = lr_network, 
  sig_network = sig_network, 
  gr_network = new_gr_network, 
  source_weights_df = new_source_weights_df %>%
    dplyr::select(source, avg_weight) %>%
    rename(weight=avg_weight))

# downweigh the importance of signaling and gene regulatory hubs - use the optimized parameters of this
weighted_networks <- apply_hub_corrections(
  weighted_networks = weighted_networks,
  lr_sig_hub = hyperparameter_list %>%
    filter(parameter == "lr_sig_hub") %>%
    pull(avg_weight),
  gr_hub = hyperparameter_list %>%
    filter(parameter == "gr_hub") %>%
    pull(avg_weight)) 

# Infer ligand-target regulatory potential scores based on the weighted integrated networks
ligands <- list("HGF")

# only one row is adapted, so should be corrected
ligand_target_matrix_HGF_24_ATAC_RNA <- construct_ligand_target_matrix(
  weighted_networks = weighted_networks, 
  ligands = ligands, algorithm = "PPR",
  damping_factor = hyperparameter_list %>% 
    filter(parameter == "damping_factor") %>% 
    pull(avg_weight),
  ltf_cutoff = hyperparameter_list %>% 
    filter(parameter == "ltf_cutoff") %>% 
    pull(avg_weight))

# there are more genes added, we should filter that since the dimensions no longer match, how can we select them in the right order without changing the format? Still need to address this.
ligand_target_matrix_HGF_24_ATAC_RNA <-
 subset(ligand_target_matrix_HGF_24_ATAC_RNA, rownames(ligand_target_matrix_HGF_24_ATAC_RNA) %in% rownames(ligand_target_matrix))

# Plot the differences in RP
plot(ligand_target_matrix[,"HGF"], ligand_target_matrix_HGF_24_ATAC_RNA[,"HGF"],
     xlab = "NicheNet default regulatory potentials",
     ylab = "Regulatory potentials with ATAC+RNA",
     main = "Regulatory potentials HGF 24h")

##OSM
GRN_OSM_24_ATAC_RNA <- read.delim2("C:/Users/klara/MaStat/Thesis/GithubThesis/Thesis_NicheNet/Ananse_networks/GRN/LINCS_MCF10A_OSM_24_grn.txt", header = TRUE, sep = ";")

# adding our GRN as a new source to the optimized weights
new_gr_network <- gr_network %>% bind_rows(GRN_OSM_24_ATAC_RNA)

new_network_weights_df <- tibble(
  source = "LINCS_MCF10A_OSM_24", 
  avg_weight = 1, 
  median_weight = 1)

new_source_weights_df <- 
  optimized_source_weights_df %>%
  bind_rows(new_network_weights_df)

# aggregate the individual data sources in a weighted manner to obtain a weighted integrated signaling network
weighted_networks <- construct_weighted_networks(
  lr_network = lr_network, 
  sig_network = sig_network, 
  gr_network = new_gr_network, 
  source_weights_df = new_source_weights_df %>%
    dplyr::select(source, avg_weight) %>%
    rename(weight = avg_weight))

# downweigh the importance of signaling and gene regulatory hubs - use the optimized parameters of this
weighted_networks <- apply_hub_corrections(
  weighted_networks = weighted_networks,
  lr_sig_hub = hyperparameter_list %>%
    filter(parameter == "lr_sig_hub") %>%
    pull(avg_weight),
  gr_hub = hyperparameter_list %>%
    filter(parameter == "gr_hub") %>%
    pull(avg_weight)) 

# Infer ligand-target regulatory potential scores based on the weighted integrated networks
ligands <- list("OSM")

# only one row is adapted, so should be corrected
ligand_target_matrix_OSM_24_ATAC_RNA <- construct_ligand_target_matrix(
  weighted_networks = weighted_networks, 
  ligands = ligands, algorithm = "PPR",
  damping_factor = hyperparameter_list %>% 
    filter(parameter == "damping_factor") %>% 
    pull(avg_weight),
  ltf_cutoff = hyperparameter_list %>% 
    filter(parameter == "ltf_cutoff") %>% 
    pull(avg_weight))

# there are more genes added, we should filter that since the dimensions no longer match, how can we select them in the right order without changing the format? Still need to address this.
ligand_target_matrix_OSM_24_ATAC_RNA <-
 subset(ligand_target_matrix_OSM_24_ATAC_RNA, rownames(ligand_target_matrix_OSM_24_ATAC_RNA) %in% rownames(ligand_target_matrix))

# Plot the differences in RP
plot(ligand_target_matrix[,"OSM"], ligand_target_matrix_OSM_24_ATAC_RNA[,"OSM"],
     xlab = "NicheNet default regulatory potentials",
     ylab = "Regulatory potentials with ATAC+RNA",
     main = "Regulatory potentials OSM 24h")

##EGF
GRN_EGF_24_ATAC_RNA <- read.delim2("C:/Users/klara/MaStat/Thesis/GithubThesis/Thesis_NicheNet/Ananse_networks/GRN/LINCS_MCF10A_EGF_24_grn.txt", header = TRUE, sep = ";")

# adding our GRN as a new source to the optimized weights
new_gr_network <- gr_network %>% bind_rows(GRN_EGF_24_ATAC_RNA)

new_network_weights_df <- tibble(
  source = "LINCS_MCF10A_EGF_24", 
  avg_weight = 1, 
  median_weight = 1)

new_source_weights_df <- 
  optimized_source_weights_df %>%
  bind_rows(new_network_weights_df)

# aggregate the individual data sources in a weighted manner to obtain a weighted integrated signaling network
weighted_networks <- construct_weighted_networks(
  lr_network = lr_network, 
  sig_network = sig_network, 
  gr_network = new_gr_network, 
  source_weights_df = new_source_weights_df %>%
    dplyr::select(source, avg_weight) %>%
    rename(weight=avg_weight))

# downweigh the importance of signaling and gene regulatory hubs - use the optimized parameters of this
weighted_networks <- apply_hub_corrections(
  weighted_networks = weighted_networks,
  lr_sig_hub = hyperparameter_list %>%
    filter(parameter == "lr_sig_hub") %>%
    pull(avg_weight),
  gr_hub = hyperparameter_list %>%
    filter(parameter == "gr_hub") %>%
    pull(avg_weight)) 

# Infer ligand-target regulatory potential scores based on the weighted integrated networks
ligands <- list("EGF")

# only one row is adapted, so should be corrected
ligand_target_matrix_EGF_24_ATAC_RNA <- construct_ligand_target_matrix(
  weighted_networks = weighted_networks, 
  ligands = ligands, algorithm = "PPR",
  damping_factor = hyperparameter_list %>% 
    filter(parameter == "damping_factor") %>% 
    pull(avg_weight),
  ltf_cutoff = hyperparameter_list %>% 
    filter(parameter == "ltf_cutoff") %>% 
    pull(avg_weight))

# there are more genes added, we should filter that since the dimensions no longer match, how can we select them in the right order without changing the format? Still need to address this.
ligand_target_matrix_EGF_24_ATAC_RNA <-
 subset(ligand_target_matrix_EGF_24_ATAC_RNA, rownames(ligand_target_matrix_EGF_24_ATAC_RNA) %in% rownames(ligand_target_matrix))

# Plot the differences in RP
plot(ligand_target_matrix[,"EGF"], ligand_target_matrix_EGF_24_ATAC_RNA[,"EGF"],
     xlab = "NicheNet default regulatory potentials",
     ylab = "Regulatory potentials with ATAC+RNA",
     main = "Regulatory potentials EGF 24h")

# Make one ligand target matrix
ligand_target_matrix_ATAC_RNA <- cbind(ligand_target_matrix_EGF_24_ATAC_RNA, ligand_target_matrix_HGF_24_ATAC_RNA, ligand_target_matrix_OSM_24_ATAC_RNA)

diff_rows_EGF_ATAC_RNA <- which(ligand_target_matrix[, "EGF"] != ligand_target_matrix_EGF_24_ATAC_RNA[, 1])
diff_rows_HGF_ATAC_RNA <- which(ligand_target_matrix[, "HGF"] != ligand_target_matrix_HGF_24_ATAC_RNA[, 1])
diff_rows_OSM_ATAC_RNA <- which(ligand_target_matrix[, "OSM"] != ligand_target_matrix_OSM_24_ATAC_RNA[, 1])

diff_rows_EGF_ATAC_RNA <- as.data.frame(diff_rows_EGF_ATAC_RNA)
diff_rows_HGF_ATAC_RNA <- as.data.frame(diff_rows_HGF_ATAC_RNA)
diff_rows_OSM_ATAC_RNA <- as.data.frame(diff_rows_OSM_ATAC_RNA)

dim(diff_rows_EGF_ATAC_RNA)
dim(diff_rows_HGF_ATAC_RNA)
dim(diff_rows_OSM_ATAC_RNA)

genes_EGF_ATAC_RNA <- as.data.frame(as.character(rownames(diff_rows_EGF_ATAC_RNA)))
colnames(genes_EGF_ATAC_RNA) <- "gene"
genes_HGF_ATAC_RNA <- as.data.frame(as.character(rownames(diff_rows_HGF_ATAC_RNA)))
colnames(genes_HGF_ATAC_RNA) <- "gene"
genes_OSM_ATAC_RNA <- as.data.frame(as.character(rownames(diff_rows_OSM_ATAC_RNA)))
colnames(genes_OSM_ATAC_RNA) <- "gene"

DE_genes_EGF <- as.data.frame(as.character(DE_info$EGF$external_gene_name))
colnames(DE_genes_EGF) <- "gene"
DE_genes_HGF <- as.data.frame(DE_info$HGF$external_gene_name)
colnames(DE_genes_HGF) <- "gene"
DE_genes_OSM <- as.data.frame(DE_info$OSM$external_gene_name)
colnames(DE_genes_OSM) <- "gene"

overlapping_genes_EGF_ATAC_RNA <- intersect(genes_EGF_ATAC_RNA, DE_genes_EGF)
dim(overlapping_genes_EGF_ATAC_RNA)
overlapping_genes_HGF_ATAC_RNA <- intersect(genes_HGF_ATAC_RNA, DE_genes_HGF)
dim(overlapping_genes_HGF_ATAC_RNA)
overlapping_genes_OSM_ATAC_RNA <- intersect(genes_OSM_ATAC_RNA, DE_genes_OSM)
dim(overlapping_genes_OSM_ATAC_RNA)
```


## Construct ligand_target_matrix using only ATAC
```{r ligand-target-matrix construction only ATAC, echo=TRUE}
# NicheNet default networks
lr_network <- readRDS(url("https://zenodo.org/record/7074291/files/lr_network_human_21122021.rds"))
sig_network <- readRDS(url("https://zenodo.org/record/7074291/files/signaling_network_human_21122021.rds"))
gr_network <- readRDS(url("https://zenodo.org/record/7074291/files/gr_network_human_21122021.rds"))

##HGF
GRN_HGF_24_ATAC <- read.delim2("C:/Users/klara/MaStat/Thesis/GithubThesis/Thesis_NicheNet/Ananse_networks/GRN/LINCS_MCF10A_HGF_24_NO_RNAseq_grn.txt", header = TRUE, sep = ";")

# adding our GRN as a new source to the optimized weights
new_gr_network <- gr_network %>% bind_rows(GRN_HGF_24_ATAC)

new_network_weights_df <- tibble(
  source = "LINCS_MCF10A_HGF_24_NO_RNAseq", 
  avg_weight = 1, 
  median_weight = 1)

new_source_weights_df <- 
  optimized_source_weights_df %>%
  bind_rows(new_network_weights_df)

# aggregate the individual data sources in a weighted manner to obtain a weighted integrated signaling network
weighted_networks <- construct_weighted_networks(
  lr_network = lr_network, 
  sig_network = sig_network, 
  gr_network = new_gr_network, 
  source_weights_df = new_source_weights_df %>%
    dplyr::select(source, avg_weight) %>%
    rename(weight=avg_weight))

# downweigh the importance of signaling and gene regulatory hubs - use the optimized parameters of this
weighted_networks <- apply_hub_corrections(
  weighted_networks = weighted_networks,
  lr_sig_hub = hyperparameter_list %>%
    filter(parameter == "lr_sig_hub") %>%
    pull(avg_weight),
  gr_hub = hyperparameter_list %>%
    filter(parameter == "gr_hub") %>%
    pull(avg_weight)) 

# Infer ligand-target regulatory potential scores based on the weighted integrated networks
ligands <- list("HGF")

# only one row is adapted, so correct for this
ligand_target_matrix_HGF_24_ATAC <- construct_ligand_target_matrix(
  weighted_networks = weighted_networks, 
  ligands = ligands, algorithm = "PPR",
  damping_factor = hyperparameter_list %>% 
    filter(parameter == "damping_factor") %>% 
    pull(avg_weight),
  ltf_cutoff = hyperparameter_list %>% 
    filter(parameter == "ltf_cutoff") %>% 
    pull(avg_weight))

# there are more genes added, we should filter that since the dimensions no longer match, how can we select them in the right order without changing the format? Still need to address this.
ligand_target_matrix_HGF_24_ATAC <-
 subset(ligand_target_matrix_HGF_24_ATAC, rownames(ligand_target_matrix_HGF_24_ATAC) %in% rownames(ligand_target_matrix))

# Plot the differences in RP
plot(ligand_target_matrix[,"HGF"], ligand_target_matrix_HGF_24_ATAC[,"HGF"],
     xlab = "NicheNet default regulatory potentials",
     ylab = "Regulatory potentials with ATAC",
     main = "Regulatory potentials HGF 24h")

##OSM
GRN_OSM_24_ATAC <- read.delim2("C:/Users/klara/MaStat/Thesis/GithubThesis/Thesis_NicheNet/Ananse_networks/GRN/LINCS_MCF10A_OSM_24_NO_RNAseq_grn.txt", header = TRUE, sep = ";")

# adding our GRN as a new source to the optimized weights
new_gr_network <- gr_network %>% bind_rows(GRN_OSM_24_ATAC)

new_network_weights_df <- tibble(
  source = "LINCS_MCF10A_OSM_24_NO_RNAseq", 
  avg_weight = 1, 
  median_weight = 1)

new_source_weights_df <- 
  optimized_source_weights_df %>%
  bind_rows(new_network_weights_df)

# aggregate the individual data sources in a weighted manner to obtain a weighted integrated signaling network
weighted_networks <- construct_weighted_networks(
  lr_network = lr_network, 
  sig_network = sig_network, 
  gr_network = new_gr_network, 
  source_weights_df = new_source_weights_df %>%
    dplyr::select(source, avg_weight) %>%
    rename(weight = avg_weight))

# downweigh the importance of signaling and gene regulatory hubs - use the optimized parameters of this
weighted_networks <- apply_hub_corrections(
  weighted_networks = weighted_networks,
  lr_sig_hub = hyperparameter_list %>%
    filter(parameter == "lr_sig_hub") %>%
    pull(avg_weight),
  gr_hub = hyperparameter_list %>%
    filter(parameter == "gr_hub") %>%
    pull(avg_weight)) 

# Infer ligand-target regulatory potential scores based on the weighted integrated networks
ligands <- list("OSM")

# only one row is adapted, so correct for this
ligand_target_matrix_OSM_24_ATAC <- construct_ligand_target_matrix(
  weighted_networks = weighted_networks, 
  ligands = ligands, algorithm = "PPR",
  damping_factor = hyperparameter_list %>% 
    filter(parameter == "damping_factor") %>% 
    pull(avg_weight),
  ltf_cutoff = hyperparameter_list %>% 
    filter(parameter == "ltf_cutoff") %>% 
    pull(avg_weight))

# there are more genes added, we should filter that since the dimensions no longer match, how can we select them in the right order without changing the format? Still need to address this.
ligand_target_matrix_OSM_24_ATAC <-
 subset(ligand_target_matrix_OSM_24_ATAC, rownames(ligand_target_matrix_OSM_24_ATAC) %in% rownames(ligand_target_matrix))

# Plot the differences in RP
plot(ligand_target_matrix[,"OSM"], ligand_target_matrix_OSM_24_ATAC[,"OSM"],
     xlab = "NicheNet default regulatory potentials",
     ylab = "Regulatory potentials with ATAC",
     main = "Regulatory potentials OSM 24h")

##EGF
GRN_EGF_24_ATAC <- read.delim2("C:/Users/klara/MaStat/Thesis/GithubThesis/Thesis_NicheNet/Ananse_networks/GRN/LINCS_MCF10A_EGF_24_NO_RNAseq_grn.txt", header = TRUE, sep = ";")

# adding our GRN as a new source to the optimized weights
new_gr_network <- gr_network %>% bind_rows(GRN_EGF_24_ATAC)

new_network_weights_df <- tibble(
  source = "LINCS_MCF10A_EGF_24_NO_RNAseq", 
  avg_weight = 1, 
  median_weight = 1)

new_source_weights_df <- 
  optimized_source_weights_df %>%
  bind_rows(new_network_weights_df)

# aggregate the individual data sources in a weighted manner to obtain a weighted integrated signaling network
weighted_networks <- construct_weighted_networks(
  lr_network = lr_network, 
  sig_network = sig_network, 
  gr_network = new_gr_network, 
  source_weights_df = new_source_weights_df %>%
    dplyr::select(source, avg_weight) %>%
    rename(weight=avg_weight))

# downweigh the importance of signaling and gene regulatory hubs - use the optimized parameters of this
weighted_networks <- apply_hub_corrections(
  weighted_networks = weighted_networks,
  lr_sig_hub = hyperparameter_list %>%
    filter(parameter == "lr_sig_hub") %>%
    pull(avg_weight),
  gr_hub = hyperparameter_list %>%
    filter(parameter == "gr_hub") %>%
    pull(avg_weight)) 

# Infer ligand-target regulatory potential scores based on the weighted integrated networks
ligands <- list("EGF")

# only one row is adapted, so correct for this
ligand_target_matrix_EGF_24_ATAC <- construct_ligand_target_matrix(
  weighted_networks = weighted_networks, 
  ligands = ligands, algorithm = "PPR",
  damping_factor = hyperparameter_list %>% 
    filter(parameter == "damping_factor") %>% 
    pull(avg_weight),
  ltf_cutoff = hyperparameter_list %>% 
    filter(parameter == "ltf_cutoff") %>% 
    pull(avg_weight))

# there are more genes added, we should filter that since the dimensions no longer match, how can we select them in the right order without changing the format? Still need to address this.
ligand_target_matrix_EGF_24_ATAC <-
 subset(ligand_target_matrix_EGF_24_ATAC, rownames(ligand_target_matrix_EGF_24_ATAC) %in% rownames(ligand_target_matrix))

# Plot the diffrences in RP
plot(ligand_target_matrix[,"EGF"], ligand_target_matrix_EGF_24_ATAC[,"EGF"],
     xlab = "NicheNet default regulatory potentials",
     ylab = "Regulatory potentials with ATAC",
     main = "Regulatory potentials EGF 24h")

# Make one ligand target matrix
ligand_target_matrix_ATAC <- cbind(ligand_target_matrix_EGF_24_ATAC, ligand_target_matrix_HGF_24_ATAC, ligand_target_matrix_OSM_24_ATAC)

diff_rows_EGF_ATAC <- which(ligand_target_matrix[, "EGF"] != ligand_target_matrix_EGF_24_ATAC[, 1])
diff_rows_HGF_ATAC <- which(ligand_target_matrix[, "HGF"] != ligand_target_matrix_HGF_24_ATAC[, 1])
diff_rows_OSM_ATAC <- which(ligand_target_matrix[, "OSM"] != ligand_target_matrix_OSM_24_ATAC[, 1])

diff_rows_EGF_ATAC <- as.data.frame(diff_rows_EGF_ATAC)
diff_rows_HGF_ATAC <- as.data.frame(diff_rows_HGF_ATAC)
diff_rows_OSM_ATAC <- as.data.frame(diff_rows_OSM_ATAC)

dim(diff_rows_EGF_ATAC)
dim(diff_rows_HGF_ATAC)
dim(diff_rows_OSM_ATAC)

genes_EGF_ATAC <- as.data.frame(as.character(rownames(diff_rows_EGF_ATAC)))
colnames(genes_EGF_ATAC) <- "gene"
genes_HGF_ATAC <- as.data.frame(as.character(rownames(diff_rows_HGF_ATAC)))
colnames(genes_HGF_ATAC) <- "gene"
genes_OSM_ATAC <- as.data.frame(as.character(rownames(diff_rows_OSM_ATAC)))
colnames(genes_OSM_ATAC) <- "gene"

DE_genes_EGF <- as.data.frame(as.character(DE_info$EGF$external_gene_name))
colnames(DE_genes_EGF) <- "gene"
DE_genes_HGF <- as.data.frame(as.character(DE_info$HGF$external_gene_name))
colnames(DE_genes_HGF) <- "gene"
DE_genes_OSM <- as.data.frame(as.character(DE_info$OSM$external_gene_name))
colnames(DE_genes_OSM) <- "gene"

overlapping_genes_EGF_ATAC <- intersect(genes_EGF_ATAC, DE_genes_EGF)
dim(overlapping_genes_EGF_ATAC)
overlapping_genes_HGF_ATAC <- intersect(genes_HGF_ATAC, DE_genes_HGF)
dim(overlapping_genes_HGF_ATAC)
overlapping_genes_OSM_ATAC <- intersect(genes_OSM_ATAC, DE_genes_OSM)
dim(overlapping_genes_OSM_ATAC)
```

```{r gene check 1, echo=TRUE, error=TRUE, warning=TRUE}
# First select the targets of the NicheNet ligand-target matrix
targets.NN <- unique(rownames(ligand_target_matrix))
length(targets.NN)

# Select the targets of the ligand-target matrices of Ananse with ATAC
targets.HGF_ATAC <- unique(GRN_HGF_24_ATAC$to) 
targets.EGF_ATAC <- unique(GRN_EGF_24_ATAC$to) 
targets.OSM_ATAC <- unique(GRN_OSM_24_ATAC$to) 

# Select the targets of the ligand-target matrices of Ananse with ATAC+RNA
targets.HGF_ATAC_RNA <- unique(GRN_HGF_24_ATAC_RNA$to) 
targets.EGF_ATAC_RNA <- unique(GRN_EGF_24_ATAC_RNA$to) 
targets.OSM_ATAC_RNA <- unique(GRN_OSM_24_ATAC_RNA$to)

# checking the overlap
length(intersect(targets.NN, targets.HGF_ATAC))
length(intersect(targets.NN, targets.EGF_ATAC))
length(intersect(targets.NN, targets.OSM_ATAC))

# checking the overlap
length(intersect(targets.NN, targets.HGF_ATAC_RNA))
length(intersect(targets.NN, targets.EGF_ATAC_RNA))
length(intersect(targets.NN, targets.OSM_ATAC_RNA))

# Check how many differentially expressed genes are in the ANANSE GRNs (ATAC)
table(DE_info$OSM$external_gene_name %in% GRN_OSM_24_ATAC$to)
table(DE_info$EGF$external_gene_name %in% GRN_EGF_24_ATAC$to)
table(DE_info$HGF$external_gene_name %in% GRN_HGF_24_ATAC$to)

# Check how many differentially expressed genes are in the ANANSE GRNs (ATAC+RNA)
table(DE_info$OSM$external_gene_name %in% GRN_OSM_24_ATAC_RNA$to)
table(DE_info$EGF$external_gene_name %in% GRN_EGF_24_ATAC_RNA$to)
table(DE_info$HGF$external_gene_name %in% GRN_HGF_24_ATAC_RNA$to)

table(GRN_OSM_24_ATAC_RNA$to %in% DE_info$OSM$external_gene_name)

# Count occurrences of each gene name in GRN_OSM_24_ATAC_RNA$to
gene_counts_OSM_ATAC_RNA <- table(GRN_OSM_24_ATAC_RNA$to)
gene_counts_EGF_ATAC_RNA <- table(GRN_EGF_24_ATAC_RNA$to)
gene_counts_HGF_ATAC_RNA <- table(GRN_HGF_24_ATAC_RNA$to)

gene_counts_OSM_ATAC <- table(GRN_OSM_24_ATAC$to)
gene_counts_EGF_ATAC <- table(GRN_EGF_24_ATAC$to)
gene_counts_HGF_ATAC <- table(GRN_HGF_24_ATAC$to)

# Extract gene names from DE_info$OSM$external_gene_name
matching_genes_OSM <- DE_info$OSM$external_gene_name
matching_genes_EGF <- DE_info$EGF$external_gene_name
matching_genes_HGF <- DE_info$HGF$external_gene_name

# Filter the counts for gene names present in DE_info$OSM$external_gene_name
matching_counts_OSM_ATAC_RNA <- gene_counts_OSM_ATAC_RNA[names(gene_counts_OSM_ATAC_RNA) %in% matching_genes_OSM]
matching_counts_EGF_ATAC_RNA <- gene_counts_EGF_ATAC_RNA[names(gene_counts_EGF_ATAC_RNA) %in% matching_genes_EGF]
matching_counts_HGF_ATAC_RNA <- gene_counts_HGF_ATAC_RNA[names(gene_counts_HGF_ATAC_RNA) %in% matching_genes_HGF]

matching_counts_OSM_ATAC <- gene_counts_OSM_ATAC[names(gene_counts_OSM_ATAC) %in% matching_genes_OSM]
matching_counts_EGF_ATAC <- gene_counts_EGF_ATAC[names(gene_counts_EGF_ATAC) %in% matching_genes_EGF]
matching_counts_HGF_ATAC <- gene_counts_HGF_ATAC[names(gene_counts_HGF_ATAC) %in% matching_genes_HGF]

# Total count of rows corresponding to genes in DE_info$OSM$external_gene_name
total_matching_rows_OSM_ATAC_RNA <- sum(matching_counts_OSM_ATAC_RNA)
total_matching_rows_EGF_ATAC_RNA <- sum(matching_counts_EGF_ATAC_RNA) 
total_matching_rows_HGF_ATAC_RNA <- sum(matching_counts_HGF_ATAC_RNA)

total_matching_rows_OSM_ATAC <- sum(matching_counts_OSM_ATAC)
total_matching_rows_EGF_ATAC <- sum(matching_counts_EGF_ATAC)
total_matching_rows_HGF_ATAC <- sum(matching_counts_HGF_ATAC)

print(paste("ATAC+RNA OSM: links in GRN with DE genes:", as.character(total_matching_rows_OSM_ATAC_RNA), "links without DE genes:", as.character(length(GRN_OSM_24_ATAC_RNA$to) - total_matching_rows_OSM_ATAC_RNA), "percentage:", as.character(total_matching_rows_OSM_ATAC_RNA/length(GRN_OSM_24_ATAC_RNA$to))))

print(paste("ATAC+RNA EGF: links in GRN with DE genes:", as.character(total_matching_rows_EGF_ATAC_RNA), "links without DE genes:", as.character(length(GRN_EGF_24_ATAC_RNA$to) - total_matching_rows_EGF_ATAC_RNA), "percentage:", as.character(total_matching_rows_EGF_ATAC_RNA/length(GRN_EGF_24_ATAC_RNA$to))))

print(paste("ATAC+RNA HGF: links in GRN with DE genes:", as.character(total_matching_rows_HGF_ATAC_RNA), "links without DE genes:", as.character(length(GRN_HGF_24_ATAC_RNA$to) - total_matching_rows_HGF_ATAC_RNA), "percentage:", as.character(total_matching_rows_HGF_ATAC_RNA/length(GRN_HGF_24_ATAC_RNA$to))))

print(paste("ATAC OSM: links in GRN with DE genes:", as.character(total_matching_rows_OSM_ATAC), "links without DE genes:", as.character(length(GRN_OSM_24_ATAC$to) - total_matching_rows_OSM_ATAC), "percentage:", as.character(total_matching_rows_OSM_ATAC/length(GRN_OSM_24_ATAC$to))))

print(paste("ATAC EGF: links in GRN with DE genes:", as.character(total_matching_rows_EGF_ATAC), "links without DE genes:", as.character(length(GRN_EGF_24_ATAC$to) - total_matching_rows_EGF_ATAC), "percentage:", as.character(total_matching_rows_EGF_ATAC/length(GRN_EGF_24_ATAC$to))))

print(paste("ATAC HGF: links in GRN with DE genes:", as.character(total_matching_rows_HGF_ATAC), "links without DE genes:", as.character(length(GRN_HGF_24_ATAC$to) - total_matching_rows_HGF_ATAC), "percentage:", as.character(total_matching_rows_HGF_ATAC/length(GRN_HGF_24_ATAC$to))))


#IS THIS NEEDED?
length(unique(GRN_OSM_24_ATAC$to)) - (table(DE_info$OSM$external_gene_name %in% GRN_OSM_24_ATAC$to)["TRUE"])
length(unique(GRN_EGF_24_ATAC$to)) - (table(DE_info$EGF$external_gene_name %in% GRN_EGF_24_ATAC$to)["TRUE"])
length(unique(GRN_HGF_24_ATAC$to)) - (table(DE_info$HGF$external_gene_name %in% GRN_HGF_24_ATAC$to)["TRUE"])
```

```{r gene check 2, echo=TRUE, error=TRUE, warning=TRUE}
# Obtain pointer and download RNA gene annotations
syn18779231 <- synGet(entity = 'syn18779231') 
filepath_annotations <- syn18779231$path
rna.annot <- read.csv(filepath_annotations)
targets.RNA <- unique(rna.annot$hgnc_symbol)

# how many unique human genes do we have?
length(targets.RNA) #37263

# checking the overlap
length(intersect(targets.NN, targets.RNA)) # also reasonable 27618
length(intersect(targets.HGF_ATAC, targets.RNA)) # very high 5041
length(intersect(targets.EGF_ATAC, targets.RNA)) # very high 5013
length(intersect(targets.OSM_ATAC, targets.RNA)) # very high 4948

length(intersect(targets.HGF_ATAC_RNA, targets.RNA)) # very high 3246
length(intersect(targets.EGF_ATAC_RNA, targets.RNA)) # very high 4146
length(intersect(targets.OSM_ATAC_RNA, targets.RNA)) # very high 3770

length(intersect(intersect(targets.NN, targets.HGF_ATAC), targets.RNA)) # very reasonable 3359
length(intersect(intersect(targets.NN, targets.EGF_ATAC), targets.RNA)) # very reasonable 3352
length(intersect(intersect(targets.NN, targets.OSM_ATAC), targets.RNA)) # very reasonable 3317

length(intersect(intersect(targets.NN, targets.HGF_ATAC_RNA), targets.RNA)) # very reasonable 2303
length(intersect(intersect(targets.NN, targets.EGF_ATAC_RNA), targets.RNA)) # very reasonable 2788
length(intersect(intersect(targets.NN, targets.OSM_ATAC_RNA), targets.RNA)) # very reasonable 2575
```

## Evaluation NicheNet
# Transcriptional response prediction evaluation
```{r model evaluation using ATAC and RNA, echo=TRUE, error=TRUE, warning=TRUE}
# Extract the gene expression dataset for EGF, HGF and OSM 
expression_settings_validation_DE_one <- list(designs[[1]]$expression_settings_validation_DE$EGF,
                                designs[[1]]$expression_settings_validation_DE$HGF,
                                designs[[1]]$expression_settings_validation_DE$OSM)

# Accordingly name the gene expression dataset
names_one <- c("EGF", "HGF", "OSM")
names(expression_settings_validation_DE_one) <- names_one

if(length(expression_settings_validation_DE_one) > 0){
  # Step 1: convert expression datasets to the required format to perform target gene prediction
  settings_one = expression_settings_validation_DE_one %>% lapply(convert_expression_settings_evaluation)
  settings_one = settings_one %>% discard(~length(.$from) > 1)
  designs[[1]]$settings_one <- settings_one
  
  # Step 2: calculate the target gene prediction performances, both for ATAC and ATAC+RNA
  performances_ATAC_RNA = settings_one %>% lapply(evaluate_target_prediction, ligand_target_matrix_ATAC_RNA) %>% bind_rows()
  
  performances_ATAC = settings_one %>% lapply(evaluate_target_prediction, ligand_target_matrix_ATAC) %>% bind_rows() 
    
  # Visualize some classification evaluation metrics showing the target gene prediction performance
  performances_ATAC_RNA = performances_ATAC_RNA %>% dplyr::select(-aupr, -auc_iregulon, -pearson_log_pval, -spearman_log_pval, -sensitivity_roc, -specificity_roc) %>% gather(key = scorename, value = scorevalue, auroc:spearman)
  
  performances_ATAC = performances_ATAC %>% dplyr::select(-aupr, -auc_iregulon, -pearson_log_pval, -spearman_log_pval, -sensitivity_roc, -specificity_roc) %>% gather(key = scorename, value = scorevalue, auroc:spearman)
  
  # Add scorelabels
  scorelabels = c(auroc = "AUROC", aupr_corrected = "AUPR (corrected)", auc_iregulon_corrected = "AUC-iRegulon (corrected)", 
                  pearson = "Pearson correlation", spearman = "Spearman's rank correlation", 
                  mean_rank_GST_log_pval = "Mean-rank gene-set enrichment")
  
  # Add a random score, to see if the dataset performs better than a random dataset
  scorerandom = c(auroc = 0.5, aupr_corrected = 0, auc_iregulon_corrected = 0, pearson = 0, spearman = 0, mean_rank_GST_log_pval = 0) %>% data.frame(scorevalue = .) %>% rownames_to_column("scorename")
  
  print(performances_ATAC_RNA %>%
          mutate(model = "NicheNet") %>%
          ggplot() +
          geom_violin(aes(model, scorevalue, group = model, fill = model)) +
          geom_boxplot(aes(model, scorevalue, group = model), width = 0.05) +
          scale_y_continuous("Score target prediction") +
          facet_wrap(~scorename, scales = "free", labeller = as_labeller(scorelabels)) +
          geom_hline(aes(yintercept = scorevalue), data = scorerandom, linetype = 2, color = "red") +
          theme_bw() +
          ggtitle("24h with ANANSE data ATAC+RNA")
  )
    
  print(performances_ATAC %>%
          mutate(model = "NicheNet") %>%
          ggplot() +
          geom_violin(aes(model, scorevalue, group = model, fill = model)) +
          geom_boxplot(aes(model, scorevalue, group = model), width = 0.05) +
          scale_y_continuous("Score target prediction") +
          facet_wrap(~scorename, scales = "free", labeller = as_labeller(scorelabels)) +
          geom_hline(aes(yintercept = scorevalue), data = scorerandom, linetype = 2, color = "red") +
          theme_bw() +
          ggtitle("24h with ANANSE data ATAC")
  )
  
  # Visualize some classification evaluation metrics showing the target gene prediction performance
  # Add nice labels to your measures
  scorelabels = c(auroc = "AUROC", aupr_corrected = "AUPR (corrected)",
                  pearson = "Pearson correlation", spearman = "Spearman's rank correlation")
  
  # Add a random score, to see if the dataset performs better than a random dataset
  scorerandom = c(auroc = 0.5, aupr_corrected = 0, pearson = 0, spearman = 0) %>% data.frame(scorevalue = .) %>% rownames_to_column("scorename")
  
  # Manually adapt the colors
  my_colors <- c("red", "blue", "green")

  # Visualize in boxplots, dots per ligand
  p_n_ATAC_RNA <- performances_ATAC_RNA %>%
  mutate(model = "NicheNet v2") %>%
  ggplot() +
  geom_boxplot(aes(model, scorevalue, group = model), width = 0.5, outlier.shape = NA) +
  geom_point(aes(model, scorevalue, color = setting), position = position_jitter(width = 0.2), alpha = 0.7) +
  scale_color_manual(values = my_colors) + 
  scale_y_continuous("Score target prediction") +
  facet_wrap(~scorename, scales = "free", labeller = labeller(scorename = scorelabels)) + 
  geom_hline(aes(yintercept = scorevalue), data = scorerandom, linetype = 2, color = "red") +
  theme_bw() +
  labs(title = "Transcriptional response prediction evaluation 24h including ATAC+RNA", color = "", x = "") +  
  theme(plot.title = element_text(size = 15, hjust = 0.5, face = "bold"), strip.text = element_text(face = "bold"))
  
  #Save the plot for later use
  ggsave("p_n_24_ATAC_RNA.png", plot = p_n_ATAC_RNA)
  
  # Visualize some classification evaluation metrics showing the target gene prediction performance
  # Add nice labels to your measures
  scorelabels = c(auroc = "AUROC", aupr_corrected = "AUPR (corrected)",
                  pearson = "Pearson correlation", spearman = "Spearman's rank correlation")
  
  # Add a random score, to see if the dataset performs better than a random dataset
  scorerandom = c(auroc = 0.5, aupr_corrected = 0, pearson = 0, spearman = 0) %>% data.frame(scorevalue = .) %>% rownames_to_column("scorename")
  
  # Manually adapt the colors
  my_colors <- c("red", "blue", "green")

  # Visualize in boxplots, dots per ligand
  p_n_ATAC <- performances_ATAC %>%
  mutate(model = "NicheNet v2") %>%
  ggplot() +
  geom_boxplot(aes(model, scorevalue, group = model), width = 0.5, outlier.shape = NA) +
  geom_point(aes(model, scorevalue, color = setting), position = position_jitter(width = 0.2), alpha = 0.7) +
  scale_color_manual(values = my_colors) + 
  scale_y_continuous("Score target prediction") +
  facet_wrap(~scorename, scales = "free", labeller = labeller(scorename = scorelabels)) + 
  geom_hline(aes(yintercept = scorevalue), data = scorerandom, linetype = 2, color = "red") +
  theme_bw() +
  labs(title = "Transcriptional response prediction evaluation 24h including ATAC", color = "", x = "") +  
  theme(plot.title = element_text(size = 15, hjust = 0.5, face = "bold"), strip.text = element_text(face = "bold"))
  
  #Save the plot for later use
  ggsave("p_n_24_ATAC.png", plot = p_n_ATAC_RNA)

}
```

# ligand activity prediction evaluation

```{r ligand activity prediction evaluation, echo=TRUE, error=TRUE}
# Extract the correct expression dataset
designs[[1]]$settings_one <- settings_one
names(settings_one) <- names_one
    
if(length(settings_one) > 0){
  # convert expression datasets to correct format for ligand activity prediction
  all_ligands = settings_one %>% extract_ligands_from_settings(combination = FALSE) %>% unlist()
  settings_ligand_prediction_one = settings_one %>% convert_settings_ligand_prediction(all_ligands = all_ligands, validation = TRUE)
    
   # infer ligand importances: for all ligands of interest, we assess how well a ligand explains the differential expression in a specific dataset (and we do this for all datasets).
   ligand_importances_ATAC_RNA = settings_ligand_prediction_one %>% lapply(get_single_ligand_importances, ligand_target_matrix_ATAC_RNA) %>% bind_rows()
   ligand_importances_ATAC_RNA <- do.call(data.frame, lapply(ligand_importances_ATAC_RNA, function(value) replace(value, is.infinite(value), 100000)))
   
   ligand_importances_ATAC = settings_ligand_prediction_one %>% lapply(get_single_ligand_importances, ligand_target_matrix_ATAC) %>% bind_rows()
   ligand_importances_ATAC <- do.call(data.frame, lapply(ligand_importances_ATAC, function(value) replace(value, is.infinite(value), 100000)))
    
    # Look at predictive performance of single/individual importance measures to predict ligand activity: of all ligands tested, the ligand that is truly active in a dataset should get the highest activity score (i.e. best target gene prediction performance)
   evaluation_ligand_prediction_ATAC_RNA = ligand_importances_one$setting %>% unique() %>% lapply(function(x){x}) %>% 
    lapply(wrapper_evaluate_single_importances_ligand_prediction, ligand_importances_ATAC_RNA) %>%
    bind_rows() %>% inner_join(ligand_importances_ATAC_RNA %>% distinct(setting, ligand))
   
   evaluation_ligand_prediction_ATAC = ligand_importances_one$setting %>% unique() %>% lapply(function(x){x}) %>% 
    lapply(wrapper_evaluate_single_importances_ligand_prediction, ligand_importances_ATAC) %>%
    bind_rows() %>% inner_join(ligand_importances_ATAC %>% distinct(setting, ligand))
   
  # Visualize some classification evaluation metrics showing the ligand activity prediction performance
  evaluation_ligand_prediction_ATAC_RNA = evaluation_ligand_prediction_ATAC_RNA %>% dplyr::select(-aupr, -sensitivity_roc, -specificity_roc,
                                                                                        -pearson, -spearman,
                                                                                        -mean_rank_GST_log_pval) %>% 
    gather(key = scorename, value = scorevalue, auroc:aupr_corrected)
  scorelabels = c(auroc = "AUROC", aupr_corrected = "AUPR (corrected)")
  scorerandom = c(auroc = 0.5, aupr_corrected = 0) %>% data.frame(scorevalue = .) %>% rownames_to_column("scorename")
  
  evaluation_ligand_prediction_ATAC = evaluation_ligand_prediction_ATAC %>% dplyr::select(-aupr, -sensitivity_roc, -specificity_roc,
                                                                                        -pearson, -spearman,
                                                                                        -mean_rank_GST_log_pval) %>% 
    gather(key = scorename, value = scorevalue, auroc:aupr_corrected)
  scorelabels = c(auroc = "AUROC", aupr_corrected = "AUPR (corrected)")
  scorerandom = c(auroc = 0.5, aupr_corrected = 0) %>% data.frame(scorevalue = .) %>% rownames_to_column("scorename")
    
  print(evaluation_ligand_prediction_ATAC_RNA %>%
     filter(importance_measure %in% c("auroc", "aupr_corrected", "mean_rank_GST_log_pval", "auc_iregulon_corrected", "pearson", "spearman")) %>%
    ggplot() +
    geom_violin(aes(importance_measure, scorevalue, group = importance_measure, fill = importance_measure)) +
    geom_boxplot(aes(importance_measure, scorevalue, group = importance_measure), width = 0.1) +
    scale_y_continuous("Evaluation ligand activity prediction") +
    scale_x_discrete("Ligand activity measure") +
    facet_wrap(~scorename, scales = "free", labeller = as_labeller(scorelabels)) +
    geom_hline(aes(yintercept = scorevalue), data = scorerandom, linetype = 2, color = "red") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90))) +
    ggtitle("24h with ANANSE data ATAC + RNA")
  
  print(evaluation_ligand_prediction_ATAC %>%
     filter(importance_measure %in% c("auroc", "aupr_corrected", "mean_rank_GST_log_pval", "auc_iregulon_corrected", "pearson", "spearman")) %>%
    ggplot() +
    geom_violin(aes(importance_measure, scorevalue, group = importance_measure, fill = importance_measure)) +
    geom_boxplot(aes(importance_measure, scorevalue, group = importance_measure), width = 0.1) +
    scale_y_continuous("Evaluation ligand activity prediction") +
    scale_x_discrete("Ligand activity measure") +
    facet_wrap(~scorename, scales = "free", labeller = as_labeller(scorelabels)) +
    geom_hline(aes(yintercept = scorevalue), data = scorerandom, linetype = 2, color = "red") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90))) +
    ggtitle("24h with ANANSE data ATAC")
}

# Visualisation thesis
  # Select metrics to visualize
  metrics <- c("auroc", "aupr_corrected", "pearson", "spearman")
  my_colors <- c("red", "blue", "green")

  # Create a list of plots for specific metrics
  plots <- lapply(metrics, function(metric) {
  ggplot(ligand_importances_ATAC_RNA, aes(x = setting, y = .data[[metric]])) +
    geom_boxplot() +
    geom_jitter(aes(color = factor(test_ligand)), width = 0.2, alpha = 0.7,  size = 3) +
    labs(title = metric,
         x = " ",
         y = metric) +
    scale_color_manual(values = my_colors) + 
    theme_minimal() +
    theme(plot.title = element_text(size = 20, hjust = 0.5),
          plot.margin = margin(1, 1, 1, 1, "cm"),
          legend.position = "none",
      axis.title = element_text(size = 18),  
      axis.text = element_text(size = 18),   
      axis.ticks = element_line(size = 1),  #
      strip.text = element_text(size = 18),
      axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
    theme(legend.title = element_blank(), legend.text = element_text(size = 20), axis.text.x = element_text(angle = 0, hjust = 0.5))
    })
  
  # Combine plots 
  combined_plots <- wrap_plots(plots, ncol = 2) +
  plot_annotation(title = "Ligand activity prediction evaluation24h ATAC+RNA", theme = theme(plot.title = element_text(size = 24, hjust = 0.5))) +
  theme(legend.position = "bottom") 
  
  combined_plots
  ggsave("ligand_activity_ATAC_RNA_24.png", plot = combined_plots, width = 13, height = 12, dpi = 300)
  
  # Visualisation thesis
  # Select metrics to visualize
  metrics <- c("auroc", "aupr_corrected", "pearson", "spearman")
  my_colors <- c("red", "blue", "green")

  # Create a list of plots for specific metrics
  plots <- lapply(metrics, function(metric) {
  ggplot(ligand_importances_ATAC, aes(x = setting, y = .data[[metric]])) +
    geom_boxplot() +
    geom_jitter(aes(color = factor(test_ligand)), width = 0.2, alpha = 0.7,  size = 3) +
    labs(title = metric,
         x = " ",
         y = metric) +
    scale_color_manual(values = my_colors) + 
    theme_minimal() +
    theme(plot.title = element_text(size = 20, hjust = 0.5),
          plot.margin = margin(1, 1, 1, 1, "cm"),
          legend.position = "none",
      axis.title = element_text(size = 18),  
      axis.text = element_text(size = 18),   
      axis.ticks = element_line(size = 1),  #
      strip.text = element_text(size = 18),
      axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
    theme(legend.title = element_blank(), legend.text = element_text(size = 20), axis.text.x = element_text(angle = 0, hjust = 0.5))
    })
  
  # Combine plots 
  combined_plots <- wrap_plots(plots, ncol = 2) +
  plot_annotation(title = "Ligand activity prediction evaluation 24h ATAC", theme = theme(plot.title = element_text(size = 24, hjust = 0.5))) +
  theme(legend.position = "bottom") 
  
  combined_plots
  ggsave("ligand_activity_ATAC_24.png", plot = combined_plots, width = 13, height = 12, dpi = 300)
```
#Statsplots NicheNet against all ATAC data
## The whole NicheNet ligand target matrix vs ATAC+RNA
```{r performances statsplots NicheNet vs ATAC+RNA, echo=TRUE, error=TRUE, fig.width=10, fig.height = 10}
performances_NicheNet["condition"] <- "NicheNet" 
performances_ATAC_RNA["condition"] <- "ATAC + RNA" 
perf_ATAC_RNA <- performances_NicheNet %>% rbind(performances_ATAC_RNA)
perf_ATAC_RNA$condition <- factor(perf_ATAC_RNA$condition)
perf_ATAC_RNA$condition <- relevel(perf_ATAC_RNA$condition, "NicheNet")

# AUROC
au_ATAC_RNA <- perf_ATAC_RNA[which(perf_ATAC_RNA$scorename == "auroc"),]
au_plot_ATAC_RNA <- ggwithinstats(
  data = au_ATAC_RNA,
  x    = condition,
  y    = scorevalue,
  xlab = " ",
  ylab = "AUROC",
  type = "np"
) +
  ggtitle("AUROC NicheNet vs ATAC+RNA 24h") +
  geom_text(aes(label = au_ATAC_RNA$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

# AUPR
aupr_ATAC_RNA <- perf_ATAC_RNA[which(perf_ATAC_RNA$scorename == "aupr_corrected"),]
aupr_plot_ATAC_RNA <- ggwithinstats(
  data = aupr_ATAC_RNA,
  x    = condition,
  y    = scorevalue,
  xlab = " ",
  ylab = "AUPR",
  type = "np"
) +
  ggtitle ("AUPR NicheNet vs ATAC+RNA 24h") +
  geom_text(aes(label = aupr_ATAC_RNA$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

# gene set enrichment
gs_ATAC_RNA <- perf_ATAC_RNA[which(perf_ATAC_RNA$scorename == "mean_rank_GST_log_pval"),]
gs_plot_ATAC_RNA <- ggwithinstats(
  data = gs_ATAC_RNA,
  x    = condition,
  y    = scorevalue,
  xlab = " ",
  ylab = "Mean rank GST log pval",
  type = "np"
) +
  ggtitle ("mean rank GST log pval NicheNet vs ATAC+RNA 24h") +
  geom_text(aes(label = gs_ATAC_RNA$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

# AUC iregulaon
iregulon_ATAC_RNA <- perf_ATAC_RNA[which(perf_ATAC_RNA$scorename == "auc_iregulon_corrected"),]
iregulon_plot_ATAC_RNA <- ggwithinstats(
  data = iregulon_ATAC_RNA,
  x    = condition,
  y    = scorevalue,
  xlab = " ",
  ylab = "AUC",
  type = "np"
) +
  ggtitle ("AUC NicheNet vs ATAC+RNA 24h") +
  geom_text(aes(label = iregulon_ATAC_RNA$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

# Spearman
sp_ATAC_RNA <- perf_ATAC_RNA[which(perf_ATAC_RNA$scorename == "spearman"),]
sp_plot_ATAC_RNA <- ggwithinstats(
  data = sp_ATAC_RNA,
  x    = condition,
  y    = scorevalue,
  xlab = " ",
  ylab = "Spearman",
  type = "np"
) +
  ggtitle ("Spearman NicheNet vs ATAC+RNA 24h") +
  geom_text(aes(label = sp_ATAC_RNA$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

# Pearson
pe_ATAC_RNA <- perf_ATAC_RNA[which(perf_ATAC_RNA$scorename == "pearson"),]
pe_plot_ATAC_RNA <- ggwithinstats(
  data = pe_ATAC_RNA,
  x    = condition,
  y    = scorevalue,
  xlab = " ",
  ylab = "Pearson",
  type = "np"
) +
  ggtitle ("Pearson NicheNet vs ATAC+RNA 24h") +
  geom_text(aes(label = pe_ATAC_RNA$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))


# Combine plots in one visualisation
combined_plot <- plot_grid(au_plot_ATAC_RNA, aupr_plot_ATAC_RNA, pe_plot_ATAC_RNA, sp_plot_ATAC_RNA)
ggsave("perf_stats_ATAC_RNA_24.png", combined_plot, width = 13, height = 11)
```


##NicheNet model vs ATAC
```{r statsplots NicheNet ligands vs ATAC 2, echo=TRUE, error=TRUE, fig.width=10, fig.height = 10}
performances_NicheNet["condition"] <- "NicheNet" 
performances_ATAC["condition"] <- "ATAC" 
perf_ATAC <- performances_NicheNet %>% rbind(performances_ATAC)
perf_ATAC$condition <- factor(perf_ATAC$condition)
perf_ATAC$condition <- relevel(perf_ATAC$condition, "NicheNet")

# AUROC
au_ATAC <- perf_ATAC[which(perf_ATAC$scorename == "auroc"),]
au_plot_ATAC <- ggwithinstats(
  data = au_ATAC,
  x    = condition,
  y    = scorevalue,
  xlab = " ",
  ylab = "AUROC",
  type = "np"
) +
  ggtitle("AUROC NicheNet vs ATAC 24h") +
  geom_text(aes(label = au_ATAC$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

# AUPR
aupr_ATAC <- perf_ATAC[which(perf_ATAC$scorename == "aupr_corrected"),]
aupr_plot_ATAC <- ggwithinstats(
  data = aupr_ATAC,
  x    = condition,
  y    = scorevalue,
  xlab = " ",
  ylab = "AUPR",
  type = "np"
) +
  ggtitle ("AUPR NicheNet vs ATAC 24h") +
  geom_text(aes(label = aupr_ATAC$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

# gene set enrichment
gs_ATAC <- perf_ATAC[which(perf_ATAC$scorename == "mean_rank_GST_log_pval"),]
gs_plot_ATAC <- ggwithinstats(
  data = gs_ATAC,
  x    = condition,
  y    = scorevalue,
  xlab = " ",
  ylab = "Mean rank GST log pval",
  type = "np"
) +
  ggtitle ("mean rank GST log pval NicheNet vs ATAC 24h") +
  geom_text(aes(label = gs_ATAC$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

# AUC iregulaon
iregulon_ATAC <- perf_ATAC[which(perf_ATAC$scorename == "auc_iregulon_corrected"),]
iregulon_plot_ATAC <- ggwithinstats(
  data = iregulon_ATAC,
  x    = condition,
  y    = scorevalue,
  xlab = " ",
  ylab = "AUC",
  type = "np"
) +
  ggtitle ("AUC NicheNet vs ATAC 24h") +
  geom_text(aes(label = iregulon_ATAC$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

# Spearman
sp_ATAC <- perf_ATAC[which(perf_ATAC$scorename == "spearman"),]
sp_plot_ATAC <- ggwithinstats(
  data = sp_ATAC,
  x    = condition,
  y    = scorevalue,
  xlab = " ",
  ylab = "Spearman",
  type = "np"
) +
  ggtitle ("Spearman NicheNet vs ATAC 24h") +
  geom_text(aes(label = sp_ATAC$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

# Pearson
pe_ATAC <- perf_ATAC[which(perf_ATAC$scorename == "pearson"),]
pe_plot_ATAC <- ggwithinstats(
  data = pe_ATAC,
  x    = condition,
  y    = scorevalue,
  xlab = " ",
  ylab = "Pearson",
  type = "np"
) +
  ggtitle ("Pearson NicheNet vs ATAC 24h") +
  geom_text(aes(label = pe_ATAC$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

# Combine plots in one visualisation
combined_plot <- plot_grid(au_plot_ATAC, aupr_plot_ATAC, pe_plot_ATAC, sp_plot_ATAC)
ggsave("perf_stats_ATAC_24.png", combined_plot, width = 13, height = 11)
```

```{r statsplots ligand acitivty prediction NicheNet ligands vs ATAC + RNA, echo=TRUE, error=TRUE, fig.width = 8, fig.height = 5}
# Add a condition to the ligand importances matrices
ligand_importances_one["condition"] <- "NicheNet"
ligand_importances_ATAC_RNA["condition"] <- "ATAC + RNA" 

# Bind the ATAC+RNA and NicheNet ligand importances
ligand_activity <- ligand_importances_one %>% rbind(ligand_importances_ATAC_RNA)

## AUPR
# Select for each ligand
EGF_aupr <- ligand_activity[which(ligand_activity$ligand == "EGF"), c("aupr", "ligand", "test_ligand", "condition")]
OSM_aupr <- ligand_activity[which(ligand_activity$ligand == "OSM"), c("aupr", "ligand", "test_ligand", "condition")]
HGF_aupr <- ligand_activity[which(ligand_activity$ligand == "HGF"), c("aupr", "ligand", "test_ligand", "condition")]

aupr_test_ligands_EGF <- EGF_aupr[,"test_ligand"]
aupr_test_ligands_OSM <- OSM_aupr[,"test_ligand"]
aupr_test_ligands_HGF <- HGF_aupr[,"test_ligand"]

plt_EGF_aupr <- ggwithinstats(
  data = EGF_aupr,
  x = condition, 
  y    = aupr,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
) +
  geom_text(aes(label = aupr_test_ligands_EGF), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  scale_x_discrete(limits = c("NicheNet", "ATAC + RNA")) +
  coord_cartesian(ylim = c(0, 0.4)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("AUPR EGF 24h") 

plt_HGF_aupr <- ggwithinstats(
  data = HGF_aupr,
  x    = condition, 
  y    = aupr,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
)+
  geom_text(aes(label = aupr_test_ligands_HGF), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +

  scale_x_discrete(limits = c("NicheNet", "ATAC + RNA")) +
  coord_cartesian(ylim = c(0, 0.4)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("AUPR HGF 24h")

plt_OSM_aupr <- ggwithinstats(
  data = OSM_aupr,
  x = condition,
  y    = aupr,
  xlab = NULL,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
) +
  geom_text(aes(label = aupr_test_ligands_OSM), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  scale_x_discrete(limits = c("NicheNet", "ATAC + RNA")) +
  coord_cartesian(ylim = c(0, 0.4)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("AUPR OSM 24h")

# Combine and save the plots
AUPR_24_ATAC_RNA <- plot_grid(plt_EGF_aupr, plt_HGF_aupr, plt_OSM_aupr, nrow = 1)
ggsave("ligand_AUPR_24_ATAC_RNA.png", AUPR_24_ATAC_RNA, height = 5, width = 10)

## AUROC
# Select for each ligand
EGF_auroc <- ligand_activity[which(ligand_activity$ligand == "EGF"), c("auroc", "ligand", "test_ligand", "condition")]
OSM_auroc <- ligand_activity[which(ligand_activity$ligand == "OSM"), c("auroc", "ligand", "test_ligand", "condition")]
HGF_auroc <- ligand_activity[which(ligand_activity$ligand == "HGF"), c("auroc", "ligand", "test_ligand", "condition")]

auroc_test_ligands_EGF <- EGF_auroc[,"test_ligand"]
auroc_test_ligands_OSM <- OSM_auroc[,"test_ligand"]
auroc_test_ligands_HGF <- HGF_auroc[,"test_ligand"]

plt_EGF_auroc <- ggwithinstats(
  data = EGF_auroc,
  x = condition, 
  y    = auroc,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
)+
  geom_text(aes(label = auroc_test_ligands_EGF), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  scale_x_discrete(limits = c("NicheNet", "ATAC + RNA")) +
  coord_cartesian(ylim = c(0.6, 0.75)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("AUROC EGF 24h") 

plt_HGF_auroc <- ggwithinstats(
  data = HGF_auroc,
  x    = condition, 
  y    = auroc,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
)+
  geom_text(aes(label = auroc_test_ligands_HGF), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +

  scale_x_discrete(limits = c("NicheNet", "ATAC + RNA")) +
  coord_cartesian(ylim = c(0.6, 0.75)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("AUROC HGF 24h")

plt_OSM_auroc <- ggwithinstats(
  data = OSM_auroc,
  x = condition,
  y    = auroc,
  xlab = NULL,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
) +
  geom_text(aes(label = auroc_test_ligands_OSM), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  scale_x_discrete(limits = c("NicheNet", "ATAC + RNA")) +
  coord_cartesian(ylim = c(0.6, 0.75)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("AUROC OSM 24h")

#Combine and save the plots
auroc_24_ATAC_RNA <- plot_grid(plt_EGF_auroc, plt_HGF_auroc, plt_OSM_auroc, nrow = 1)
ggsave("ligand_auroc_24_ATAC_RNA.png", auroc_24_ATAC_RNA, height = 5, widt = 10)

## Pearson
# Select for each ligand
EGF_pearson <- ligand_activity[which(ligand_activity$ligand == "EGF"), c("pearson", "ligand", "test_ligand", "condition")]
OSM_pearson <- ligand_activity[which(ligand_activity$ligand == "OSM"), c("pearson", "ligand", "test_ligand", "condition")]
HGF_pearson <- ligand_activity[which(ligand_activity$ligand == "HGF"), c("pearson", "ligand", "test_ligand", "condition")]

pearson_test_ligands_EGF <- EGF_pearson[,"test_ligand"]
pearson_test_ligands_OSM <- OSM_pearson[,"test_ligand"]
pearson_test_ligands_HGF <- HGF_pearson[,"test_ligand"]

plt_EGF_pearson <- ggwithinstats(
  data = EGF_pearson,
  x = condition, 
  y    = pearson,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
)+
  geom_text(aes(label = pearson_test_ligands_EGF), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  scale_x_discrete(limits = c("NicheNet", "ATAC + RNA")) +
  coord_cartesian(ylim = c(0, 0.3)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("Pearson EGF 24h") 

plt_HGF_pearson <- ggwithinstats(
  data = HGF_pearson,
  x    = condition, 
  y    = pearson,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
)+
  geom_text(aes(label = pearson_test_ligands_HGF), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +

  scale_x_discrete(limits = c("NicheNet", "ATAC + RNA")) +
  coord_cartesian(ylim = c(0, 0.3)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("Pearson HGF 24h")

plt_OSM_pearson <- ggwithinstats(
  data = OSM_pearson,
  x = condition,
  y    = pearson,
  xlab = NULL,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
)+
  geom_text(aes(label = pearson_test_ligands_OSM), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  scale_x_discrete(limits = c("NicheNet", "ATAC + RNA")) +
  coord_cartesian(ylim = c(0, 0.3)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("Pearson OSM 24h")

# Combine and save the plots
pearson_24_ATAC_RNA <- plot_grid(plt_EGF_pearson, plt_HGF_pearson, plt_OSM_pearson, nrow = 1)
ggsave("ligand_pearson_24_ATAC_RNA.png", pearson_24_ATAC_RNA, height = 5, width = 10)

## Spearman
# Select for each ligand
EGF_spearman <- ligand_activity[which(ligand_activity$ligand == "EGF"), c("spearman", "ligand", "test_ligand", "condition")]
OSM_spearman <- ligand_activity[which(ligand_activity$ligand == "OSM"), c("spearman", "ligand", "test_ligand", "condition")]
HGF_spearman <- ligand_activity[which(ligand_activity$ligand == "HGF"), c("spearman", "ligand", "test_ligand", "condition")]

spearman_test_ligands_EGF <- EGF_spearman[,"test_ligand"]
spearman_test_ligands_OSM <- OSM_spearman[,"test_ligand"]
spearman_test_ligands_HGF <- HGF_spearman[,"test_ligand"]

plt_EGF_spearman <- ggwithinstats(
  data = EGF_spearman,
  x = condition, 
  y  = spearman,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
)+
  geom_text(aes(label = spearman_test_ligands_EGF), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  scale_x_discrete(limits = c("NicheNet", "ATAC + RNA")) +
  coord_cartesian(ylim = c(0, 0.3)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("Spearman EGF 24h") 

plt_HGF_spearman <- ggwithinstats(
  data = HGF_spearman,
  x    = condition, 
  y    = spearman,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
)+
  geom_text(aes(label = spearman_test_ligands_HGF), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  scale_x_discrete(limits = c("NicheNet", "ATAC + RNA")) +
  coord_cartesian(ylim = c(0, 0.3)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("Spearman HGF 24h")

plt_OSM_spearman <- ggwithinstats(
  data = OSM_spearman,
  x = condition,
  y    = spearman,
  xlab = NULL,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
)+
  geom_text(aes(label = spearman_test_ligands_OSM), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  scale_x_discrete(limits = c("NicheNet", "ATAC + RNA")) +
  coord_cartesian(ylim = c(0, 0.3)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("Spearman OSM")

#Combine plots and save
spearman_24_ATAC_RNA <- plot_grid(plt_EGF_spearman, plt_HGF_spearman, plt_OSM_spearman, nrow = 1)
ggsave("ligand_spearman_24_ATAC_RNA.png", spearman_24_ATAC_RNA, height = 5, width = 10)
```

```{r statsplots ligand acitivty prediction NicheNet ligands vs ATAC, echo=TRUE, error=TRUE, fig.width = 8, fig.height = 5}
# Add a condition to the ligand importances matrices
ligand_importances_one["condition"] <- "NicheNet"
ligand_importances_ATAC["condition"] <- "ATAC" 

# Bind the ATAC and NicheNet ligand importances
ligand_activity <- ligand_importances_one %>% rbind(ligand_importances_ATAC)

## AUPR
# Select for each ligand
EGF_aupr <- ligand_activity[which(ligand_activity$ligand == "EGF"), c("aupr", "ligand", "test_ligand", "condition")]
OSM_aupr <- ligand_activity[which(ligand_activity$ligand == "OSM"), c("aupr", "ligand", "test_ligand", "condition")]
HGF_aupr <- ligand_activity[which(ligand_activity$ligand == "HGF"), c("aupr", "ligand", "test_ligand", "condition")]

aupr_test_ligands_EGF <- EGF_aupr[,"test_ligand"]
aupr_test_ligands_OSM <- OSM_aupr[,"test_ligand"]
aupr_test_ligands_HGF <- HGF_aupr[,"test_ligand"]

plt_EGF_aupr <- ggwithinstats(
  data = EGF_aupr,
  x = condition, 
  y    = aupr,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
)+
  geom_text(aes(label = aupr_test_ligands_EGF), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  scale_x_discrete(limits = c("NicheNet", "ATAC")) +
  coord_cartesian(ylim = c(0, 0.4)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("AUPR EGF 24h") 

plt_HGF_aupr <- ggwithinstats(
  data = HGF_aupr,
  x    = condition, 
  y    = aupr,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
)+
  geom_text(aes(label = aupr_test_ligands_HGF), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  scale_x_discrete(limits = c("NicheNet", "ATAC")) +
  coord_cartesian(ylim = c(0, 0.4)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("AUPR HGF 24h")

plt_OSM_aupr <- ggwithinstats(
  data = OSM_aupr,
  x = condition,
  y    = aupr,
  xlab = NULL,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
) +
  geom_text(aes(label = aupr_test_ligands_OSM), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  scale_x_discrete(limits = c("NicheNet", "ATAC")) +
  coord_cartesian(ylim = c(0, 0.4)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("AUPR OSM 24h")

# Combine and save the plots
AUPR_24_ATAC <- plot_grid(plt_EGF_aupr, plt_HGF_aupr, plt_OSM_aupr, nrow = 1)
ggsave("ligand_AUPR_24_ATAC.png", AUPR_24_ATAC, height = 5, width = 10)

## AUROC
# Select for each ligand
EGF_auroc <- ligand_activity[which(ligand_activity$ligand == "EGF"), c("auroc", "ligand", "test_ligand", "condition")]
EGF_auroc$condition <- factor(EGF_auroc$condition, levels = c("NicheNet", "ATAC"))
OSM_auroc <- ligand_activity[which(ligand_activity$ligand == "OSM"), c("auroc", "ligand", "test_ligand", "condition")]
OSM_auroc$condition <- factor(OSM_auroc$condition, levels = c("NicheNet", "ATAC"))
HGF_auroc <- ligand_activity[which(ligand_activity$ligand == "HGF"), c("auroc", "ligand", "test_ligand", "condition")]
HGF_auroc$condition <- factor(HGFF_auroc$condition, levels = c("NicheNet", "ATAC"))

auroc_test_ligands_EGF <- EGF_auroc[,"test_ligand"]
auroc_test_ligands_OSM <- OSM_auroc[,"test_ligand"]
auroc_test_ligands_HGF <- HGF_auroc[,"test_ligand"]

plt_EGF_auroc <- ggwithinstats(
  data = EGF_auroc,
  x = condition, 
  y    = auroc,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
) +
  geom_text(aes(label = auroc_test_ligands_EGF), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  coord_cartesian(ylim = c(0.6, 0.75)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("AUROC EGF 24h") 

plt_HGF_auroc <- ggwithinstats(
  data = HGF_auroc,
  x    = condition, 
  y    = auroc,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
) +
  geom_text(aes(label = auroc_test_ligands_HGF), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +

  scale_x_discrete(limits = c("NicheNet", "ATAC")) +
  coord_cartesian(ylim = c(0.6, 0.75)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("AUROC HGF 24h")

plt_OSM_auroc <- ggwithinstats(
  data = OSM_auroc,
  x = condition,
  y    = auroc,
  xlab = NULL,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
) +
  geom_text(aes(label = auroc_test_ligands_OSM), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  scale_x_discrete(limits = c("NicheNet", "ATAC")) +
  coord_cartesian(ylim = c(0.6, 0.75)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("AUROC OSM 24h")

#Combine and save the plots
auroc_24_ATAC <- plot_grid(plt_EGF_auroc, plt_HGF_auroc, plt_OSM_auroc, nrow = 1)
ggsave("ligand_auroc_24_ATAC.png", auroc_24_ATAC, height = 5, width = 10)

## Pearson
# Select for each ligand
EGF_pearson <- ligand_activity[which(ligand_activity$ligand == "EGF"), c("pearson", "ligand", "test_ligand", "condition")]
EGF_pearson$condition <- factor(EGF_pearson$condition, levels = c("NicheNet", "ATAC"))
OSM_pearson <- ligand_activity[which(ligand_activity$ligand == "OSM"), c("pearson", "ligand", "test_ligand", "condition")]
HGF_pearson <- ligand_activity[which(ligand_activity$ligand == "HGF"), c("pearson", "ligand", "test_ligand", "condition")]

pearson_test_ligands_EGF <- EGF_pearson[,"test_ligand"]
pearson_test_ligands_OSM <- OSM_pearson[,"test_ligand"]
pearson_test_ligands_HGF <- HGF_pearson[,"test_ligand"]

plt_EGF_pearson <- ggwithinstats(
  data = EGF_pearson,
  x = condition, 
  y    = pearson,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
) +
  geom_text(aes(label = pearson_test_ligands_EGF), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  coord_cartesian(ylim = c(0, 0.2)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("Pearson EGF 24h") 

plt_HGF_pearson <- ggwithinstats(
  data = HGF_pearson,
  x    = condition, 
  y    = pearson,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
)+
  geom_text(aes(label = pearson_test_ligands_HGF), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  scale_x_discrete(limits = c("NicheNet", "ATAC")) +
  coord_cartesian(ylim = c(0, 0.2)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("Pearson HGF 24h")

plt_OSM_pearson <- ggwithinstats(
  data = OSM_pearson,
  x = condition,
  y    = pearson,
  xlab = NULL,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
)+
  geom_text(aes(label = pearson_test_ligands_OSM), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  scale_x_discrete(limits = c("NicheNet", "ATAC")) +
  coord_cartesian(ylim = c(0, 0.2)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("Pearson OSM 24h")

# Combine and save the plots
pearson_24_ATAC <- plot_grid(plt_EGF_pearson, plt_HGF_pearson, plt_OSM_pearson, nrow = 1)
ggsave("ligand_pearson_24_ATAC.png", pearson_24_ATAC, height = 5, width = 10)

## Spearman
# Select for each ligand
EGF_spearman <- ligand_activity[which(ligand_activity$ligand == "EGF"), c("spearman", "ligand", "test_ligand", "condition")]
OSM_spearman <- ligand_activity[which(ligand_activity$ligand == "OSM"), c("spearman", "ligand", "test_ligand", "condition")]
HGF_spearman <- ligand_activity[which(ligand_activity$ligand == "HGF"), c("spearman", "ligand", "test_ligand", "condition")]

spearman_test_ligands_EGF <- EGF_spearman[,"test_ligand"]
spearman_test_ligands_OSM <- OSM_spearman[,"test_ligand"]
spearman_test_ligands_HGF <- HGF_spearman[,"test_ligand"]

plt_EGF_spearman <- ggwithinstats(
  data = EGF_spearman,
  x = condition, 
  y  = spearman,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
)+
  geom_text(aes(label = spearman_test_ligands_EGF), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  scale_x_discrete(limits = c("NicheNet", "ATAC")) +
  coord_cartesian(ylim = c(0, 0.3)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("Spearman EGF 24h") 

plt_HGF_spearman <- ggwithinstats(
  data = HGF_spearman,
  x    = condition, 
  y    = spearman,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
)+
  geom_text(aes(label = spearman_test_ligands_HGF), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  scale_x_discrete(limits = c("NicheNet", "ATAC")) +
  coord_cartesian(ylim = c(0, 0.3)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("Spearman HGF 24h")

plt_OSM_spearman <- ggwithinstats(
  data = OSM_spearman,
  x = condition,
  y    = spearman,
  xlab = NULL,
  type = "np",
  centrality.point.args = list(size = 2, color = "darkred"),
  centrality.label.args = list(size = 2.5, nudge_x = 0.4, segment.linetype = 4),
  point.args = list(size = 2, alpha = 0.5, na.rm = TRUE)
)+
  geom_text(aes(label = spearman_test_ligands_OSM), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  scale_x_discrete(limits = c("NicheNet", "ATAC")) +
  coord_cartesian(ylim = c(0, 0.3)) +
  theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle ("Spearman OSM 24h")

#Combine plots and save
spearman_24_ATAC <- plot_grid(plt_EGF_spearman, plt_HGF_spearman, plt_OSM_spearman, nrow = 1)
ggsave("ligand_spearman_24_ATAC.png", spearman_24_ATAC, height = 5, width = 10)
```


##Construction of ligand_target_matrix: Mock analysis
In GRN.R are the weighted networks and individual ligand-target-matrices constructed.
```{r ligand-target-matrix construction percentages, without RNAseq}
# The complete networks can be downloaded from Zenodo
lr_network <- readRDS(url("https://zenodo.org/record/7074291/files/lr_network_human_21122021.rds"))
sig_network <- readRDS(url("https://zenodo.org/record/7074291/files/signaling_network_human_21122021.rds"))
gr_network <- readRDS(url("https://zenodo.org/record/7074291/files/gr_network_human_21122021.rds"))

#import one ligand GRN networks
gr_network_LINCS_OSM <- read.delim("C:/Users/klara/MaStat/Thesis/GithubThesis/Thesis_NicheNet/Ananse_networks/GRN/LINCS_MCF10A_OSM_24_grn.txt", header = TRUE, sep = ";")
gr_network_LINCS_HGF <- read.delim("C:/Users/klara/MaStat/Thesis/GithubThesis/Thesis_NicheNet/Ananse_networks/GRN/LINCS_MCF10A_HGF_24_grn.txt", header = TRUE, sep = ";")
gr_network_LINCS_EGF <- read.delim("C:/Users/klara/MaStat/Thesis/GithubThesis/Thesis_NicheNet/Ananse_networks/GRN/LINCS_MCF10A_EGF_24_grn.txt", header = TRUE, sep = ";")

#only keep DE_genes in GRN
gr_network_LINCS_OSM_DE <- gr_network_LINCS_OSM[(gr_network_LINCS_OSM$to %in% table1$DE_info$OSM$external_gene_name),]
gr_network_LINCS_HGF_DE <- gr_network_LINCS_HGF[(gr_network_LINCS_HGF$to %in% table1$DE_info$HGF$external_gene_name),]
gr_network_LINCS_EGF_DE <- gr_network_LINCS_EGF[(gr_network_LINCS_EGF$to %in% table1$DE_info$EGF$external_gene_name),]

#GRN without DEgenes
gr_network_LINCS_OSM_not <- gr_network_LINCS_OSM[!(gr_network_LINCS_OSM$to %in% table1$DE_info$OSM$external_gene_name),]
gr_network_LINCS_HGF_not <- gr_network_LINCS_HGF[!(gr_network_LINCS_HGF$to %in% table1$DE_info$HGF$external_gene_name),]
gr_network_LINCS_EGF_not <- gr_network_LINCS_EGF[!(gr_network_LINCS_EGF$to %in% table1$DE_info$EGF$external_gene_name),]

#add your new source to the weighted networks in NicheNet (weight??)
new_network_weights_df_EGF <- tibble(source = "LINCS_MCF10A_EGF_24", avg_weight = 1, median_weight = 1)
new_source_weights_df_EGF <- optimized_source_weights_df %>% bind_rows(new_network_weights_df_EGF)

new_network_weights_df_HGF <- tibble(source = "LINCS_MCF10A_HGF_24", avg_weight = 1, median_weight = 1)
new_source_weights_df_HGF <- optimized_source_weights_df %>% bind_rows(new_network_weights_df_HGF)

new_network_weights_df_OSM <- tibble(source = "LINCS_MCF10A_OSM_24", avg_weight = 1, median_weight = 1)
new_source_weights_df_OSM <- optimized_source_weights_df %>% bind_rows(new_network_weights_df_OSM)

#make new GRNs with 0,25,50,75 and 100% of DE genes
all_grn_OSM <- list()
all_grn_EGF <- list()
all_grn_HGF <- list()
names_per <- c()
all_weighted_networks_EGF <- list()
all_weighted_networks_HGF <- list()
all_weighted_networks_OSM <- list()
all_ligand_target_matrix_EGF <- list()
all_ligand_target_matrix_HGF <- list()
all_ligand_target_matrix_OSM <- list()
all_new_grn <- c()
for(per in c(0, 0.25, 0.5, 0.75, 1)){
  grn_OSM_DE <- gr_network_LINCS_OSM_DE[sample(nrow(gr_network_LINCS_OSM_DE), round(nrow(gr_network_LINCS_OSM_DE)*per)), ] %>%
             rbind(gr_network_LINCS_OSM_not[sample(nrow(gr_network_LINCS_OSM_not),
                                                   round(nrow(gr_network_LINCS_OSM_DE)*(1-per))), ])
  grn_OSM <- gr_network %>% bind_rows(grn_OSM_DE)
  
  grn_EGF_DE <- gr_network_LINCS_EGF_DE[sample(nrow(gr_network_LINCS_EGF_DE), round(nrow(gr_network_LINCS_EGF_DE)*per)), ] %>%
             rbind(gr_network_LINCS_EGF_not[sample(nrow(gr_network_LINCS_EGF_not),
                                                   round(nrow(gr_network_LINCS_EGF_DE)*(1-per))), ])
  grn_EGF <- gr_network %>% bind_rows(grn_EGF_DE)
  
  grn_HGF_DE <- gr_network_LINCS_OSM_DE[sample(nrow(gr_network_LINCS_HGF_DE), round(nrow(gr_network_LINCS_HGF_DE)*per)), ] %>%
             rbind(gr_network_LINCS_HGF_not[sample(nrow(gr_network_LINCS_HGF_not),
                                                   round(nrow(gr_network_LINCS_HGF_DE)*(1-per))), ])
  grn_HGF <- gr_network %>% bind_rows(grn_HGF_DE)
  
  print("grn check")

  weighted_networks_EGF = construct_weighted_networks(lr_network = lr_network, sig_network = sig_network,
                                                  gr_network = grn_EGF, source_weights_df = new_source_weights_df_EGF %>%
                                                    dplyr::select(source, avg_weight) %>%
                                                    rename(weight = avg_weight))
  
  # downweigh the importance of signaling and gene regulatory hubs - use the optimized parameters of this
weighted_networks_EGF <- apply_hub_corrections(
  weighted_networks = weighted_networks_EGF,
  lr_sig_hub = hyperparameter_list %>%
    filter(parameter == "lr_sig_hub") %>%
    pull(avg_weight),
  gr_hub = hyperparameter_list %>%
    filter(parameter == "gr_hub") %>%
    pull(avg_weight))
  
  weighted_networks_HGF = construct_weighted_networks(lr_network = lr_network, sig_network = sig_network,
                                                  gr_network = grn_HGF, source_weights_df = new_source_weights_df_HGF %>%
                                                    dplyr::select(source, avg_weight) %>%
                                                    rename(weight = avg_weight))
  
  # downweigh the importance of signaling and gene regulatory hubs - use the optimized parameters of this
weighted_networks_HGF <- apply_hub_corrections(
  weighted_networks = weighted_networks_HGF,
  lr_sig_hub = hyperparameter_list %>%
    filter(parameter == "lr_sig_hub") %>%
    pull(avg_weight),
  gr_hub = hyperparameter_list %>%
    filter(parameter == "gr_hub") %>%
    pull(avg_weight))
  
  weighted_networks_OSM = construct_weighted_networks(lr_network = lr_network, sig_network = sig_network,
                                                  gr_network = grn_OSM, source_weights_df = new_source_weights_df_OSM %>%
                                                    dplyr::select(source, avg_weight) %>%
                                                    rename(weight = avg_weight))
  
  # downweigh the importance of signaling and gene regulatory hubs - use the optimized parameters of this
weighted_networks_OSM <- apply_hub_corrections(
  weighted_networks = weighted_networks_OSM,
  lr_sig_hub = hyperparameter_list %>%
    filter(parameter == "lr_sig_hub") %>%
    pull(avg_weight),
  gr_hub = hyperparameter_list %>%
    filter(parameter == "gr_hub") %>%
    pull(avg_weight))
  
    print("weight check")

  #construct ligand_target_matrix
  ligand_target_matrix_EGF <- construct_ligand_target_matrix(
  weighted_networks = weighted_networks_EGF, 
  ligands = list("EGF"), algorithm = "PPR",
  damping_factor = hyperparameter_list %>% 
    filter(parameter == "damping_factor") %>% 
    pull(avg_weight),
  ltf_cutoff = hyperparameter_list %>% 
    filter(parameter == "ltf_cutoff") %>% 
    pull(avg_weight))
  
  ligand_target_matrix_HGF <- construct_ligand_target_matrix(
  weighted_networks = weighted_networks_HGF, 
  ligands = list("HGF"), algorithm = "PPR",
  damping_factor = hyperparameter_list %>% 
    filter(parameter == "damping_factor") %>% 
    pull(avg_weight),
  ltf_cutoff = hyperparameter_list %>% 
    filter(parameter == "ltf_cutoff") %>% 
    pull(avg_weight))
  
  ligand_target_matrix_OSM <- construct_ligand_target_matrix(
  weighted_networks = weighted_networks_OSM, 
  ligands = list("OSM"), algorithm = "PPR",
  damping_factor = hyperparameter_list %>% 
    filter(parameter == "damping_factor") %>% 
    pull(avg_weight),
  ltf_cutoff = hyperparameter_list %>% 
    filter(parameter == "ltf_cutoff") %>% 
    pull(avg_weight))
  
  all_weighted_networks_EGF$per <- weighted_networks_EGF
  all_weighted_networks_HGF$per <- weighted_networks_HGF
  all_weighted_networks_OSM$per <- weighted_networks_OSM
  
  all_ligand_target_matrix_EGF$per <- ligand_target_matrix_EGF
  all_ligand_target_matrix_HGF$per <- ligand_target_matrix_HGF
  all_ligand_target_matrix_OSM$per <- ligand_target_matrix_OSM
  
  print("ligand_target_matrix check")

  all_grn_OSM$per <- as.data.frame(grn_OSM)
  all_grn_EGF$per <- as.data.frame(grn_EGF)
  all_grn_HGF$per <- as.data.frame(grn_HGF)
  names_per <- append(names_per, toString(per))
  names(all_grn_OSM) <- names_per
  names(all_grn_EGF) <- names_per
  names(all_grn_HGF) <- names_per
  names(all_ligand_target_matrix_EGF) <- names_per
  names(all_ligand_target_matrix_HGF) <- names_per
  names(all_ligand_target_matrix_OSM) <- names_per
  names(all_weighted_networks_EGF) <- names_per
  names(all_weighted_networks_HGF) <- names_per
  names(all_weighted_networks_OSM) <- names_per
  
}
```

```{r ligand-target-matrix construction percentages, with RNAseq}
# The complete networks can be downloaded from Zenodo
lr_network <- readRDS(url("https://zenodo.org/record/7074291/files/lr_network_human_21122021.rds"))
sig_network <- readRDS(url("https://zenodo.org/record/7074291/files/signaling_network_human_21122021.rds"))
gr_network <- readRDS(url("https://zenodo.org/record/7074291/files/gr_network_human_21122021.rds"))

#import one ligand GRN networks
gr_network_LINCS_OSM_ATAC <- read.delim("C:/Users/klara/MaStat/Thesis/GithubThesis/Thesis_NicheNet/Ananse_networks/GRN/LINCS_MCF10A_OSM_24_NO_RNAseq_grn.txt", header = TRUE, sep = ";")
gr_network_LINCS_HGF_ATAC <- read.delim("C:/Users/klara/MaStat/Thesis/GithubThesis/Thesis_NicheNet/Ananse_networks/GRN/LINCS_MCF10A_HGF_24_NO_RNAseq_grn.txt", header = TRUE, sep = ";")
gr_network_LINCS_EGF_ATAC <- read.delim("C:/Users/klara/MaStat/Thesis/GithubThesis/Thesis_NicheNet/Ananse_networks/GRN/LINCS_MCF10A_EGF_24_NO_RNAseq_grn.txt", header = TRUE, sep = ";")

#only keep DE_genes in GRN
gr_network_LINCS_OSM_DE_ATAC <- gr_network_LINCS_OSM[(gr_network_LINCS_OSM_ATAC$to %in% table1$DE_info$OSM$external_gene_name),]
gr_network_LINCS_HGF_DE_ATAC <- gr_network_LINCS_HGF[(gr_network_LINCS_HGF_ATAC$to %in% table1$DE_info$HGF$external_gene_name),]
gr_network_LINCS_EGF_DE_ATAC <- gr_network_LINCS_EGF[(gr_network_LINCS_EGF_ATAC$to %in% table1$DE_info$EGF$external_gene_name),]

#GRN without DEgenes
gr_network_LINCS_OSM_not_ATAC <- gr_network_LINCS_OSM_ATAC[!(gr_network_LINCS_OSM_ATAC$to %in% table1$DE_info$OSM$external_gene_name),]
gr_network_LINCS_HGF_not_ATAC <- gr_network_LINCS_HGF_ATAC[!(gr_network_LINCS_HGF_ATAC$to %in% table1$DE_info$HGF$external_gene_name),]
gr_network_LINCS_EGF_not_ATAC <- gr_network_LINCS_EGF_ATAC[!(gr_network_LINCS_EGF_ATAC$to %in% table1$DE_info$EGF$external_gene_name),]

#add your new source to the weighted networks in NicheNet (weight??)
new_network_weights_df_EGF_ATAC <- tibble(source = "LINCS_MCF10A_EGF_24_NO_RNAseq", avg_weight = 1, median_weight = 1)
new_source_weights_df_EGF_ATAC <- optimized_source_weights_df %>% bind_rows(new_network_weights_df_EGF_ATAC)

new_network_weights_df_HGF_ATAC <- tibble(source = "LINCS_MCF10A_HGF_24_NO_RNAseq", avg_weight = 1, median_weight = 1)
new_source_weights_df_HGF_ATAC <- optimized_source_weights_df %>% bind_rows(new_network_weights_df_HGF_ATAC)

new_network_weights_df_OSM_ATAC <- tibble(source = "LINCS_MCF10A_OSM_24_NO_RNAseq", avg_weight = 1, median_weight = 1)
new_source_weights_df_OSM_ATAC <- optimized_source_weights_df %>% bind_rows(new_network_weights_df_OSM_ATAC)

#make new GRNs with 0,25,50,75 and 100% of DE genes
all_grn_OSM_ATAC <- list()
all_grn_EGF_ATAC <- list()
all_grn_HGF_ATAC <- list()
names_per <- c()
all_weighted_networks_EGF_ATAC <- list()
all_weighted_networks_HGF_ATAC <- list()
all_weighted_networks_OSM_ATAC <- list()
all_ligand_target_matrix_EGF_ATAC <- list()
all_ligand_target_matrix_HGF_ATAC <- list()
all_ligand_target_matrix_OSM_ATAC <- list()
all_new_grn_ATAC <- c()
for(per in c(0, 0.25, 0.5, 0.75, 1)){
  grn_OSM_ATAC_DE <- gr_network_LINCS_OSM_DE_ATAC[sample(nrow(gr_network_LINCS_OSM_DE_ATAC), round(nrow(gr_network_LINCS_OSM_DE_ATAC)*per)), ] %>%
             rbind(gr_network_LINCS_OSM_not_ATAC[sample(nrow(gr_network_LINCS_OSM_not_ATAC),
                                                   round(nrow(gr_network_LINCS_OSM_DE_ATAC)*(1-per))), ])
  grn_OSM_ATAC <- gr_network %>% bind_rows(grn_OSM_ATAC_DE)
  
  grn_EGF_ATAC_DE <- gr_network_LINCS_EGF_DE_ATAC[sample(nrow(gr_network_LINCS_EGF_DE_ATAC), round(nrow(gr_network_LINCS_EGF_DE_ATAC)*per)), ] %>%
             rbind(gr_network_LINCS_EGF_not_ATAC[sample(nrow(gr_network_LINCS_EGF_not_ATAC),
                                                   round(nrow(gr_network_LINCS_EGF_DE_ATAC)*(1-per))), ])
  grn_EGF_ATAC <- gr_network %>% bind_rows(grn_EGF_ATAC_DE)
  
  grn_HGF_ATAC_DE <- gr_network_LINCS_OSM_DE_ATAC[sample(nrow(gr_network_LINCS_HGF_DE_ATAC), round(nrow(gr_network_LINCS_HGF_DE_ATAC)*per)), ] %>%
             rbind(gr_network_LINCS_HGF_not_ATAC[sample(nrow(gr_network_LINCS_HGF_not_ATAC),
                                                   round(nrow(gr_network_LINCS_HGF_DE_ATAC)*(1-per))), ])
  grn_HGF_ATAC <- gr_network %>% bind_rows(grn_HGF_ATAC_DE)
  
  print("grn check")

  weighted_networks_EGF_ATAC = construct_weighted_networks(lr_network = lr_network, sig_network = sig_network,
                                                  gr_network = grn_EGF_ATAC, source_weights_df = new_source_weights_df_EGF_ATAC %>%
                                                    dplyr::select(source, avg_weight) %>%
                                                    rename(weight = avg_weight))
  
  # downweigh the importance of signaling and gene regulatory hubs - use the optimized parameters of this
weighted_networks_EGF_ATAC <- apply_hub_corrections(
  weighted_networks = weighted_networks_EGF_ATAC,
  lr_sig_hub = hyperparameter_list %>%
    filter(parameter == "lr_sig_hub") %>%
    pull(avg_weight),
  gr_hub = hyperparameter_list %>%
    filter(parameter == "gr_hub") %>%
    pull(avg_weight))
  
  weighted_networks_HGF_ATAC = construct_weighted_networks(lr_network = lr_network, sig_network = sig_network,
                                                  gr_network = grn_HGF_ATAC, source_weights_df = new_source_weights_df_HGF_ATAC %>%
                                                    dplyr::select(source, avg_weight) %>%
                                                    rename(weight = avg_weight))
  
  # downweigh the importance of signaling and gene regulatory hubs - use the optimized parameters of this
weighted_networks_HGF_ATAC <- apply_hub_corrections(
  weighted_networks = weighted_networks_HGF_ATAC,
  lr_sig_hub = hyperparameter_list %>%
    filter(parameter == "lr_sig_hub") %>%
    pull(avg_weight),
  gr_hub = hyperparameter_list %>%
    filter(parameter == "gr_hub") %>%
    pull(avg_weight))
  
  weighted_networks_OSM_ATAC = construct_weighted_networks(lr_network = lr_network, sig_network = sig_network,
                                                  gr_network = grn_OSM_ATAC, source_weights_df = new_source_weights_df_OSM_ATAC %>%
                                                    dplyr::select(source, avg_weight) %>%
                                                    rename(weight = avg_weight))
  
  # downweigh the importance of signaling and gene regulatory hubs - use the optimized parameters of this
weighted_networks_OSM_ATAC <- apply_hub_corrections(
  weighted_networks = weighted_networks_OSM_ATAC,
  lr_sig_hub = hyperparameter_list %>%
    filter(parameter == "lr_sig_hub") %>%
    pull(avg_weight),
  gr_hub = hyperparameter_list %>%
    filter(parameter == "gr_hub") %>%
    pull(avg_weight))
  
    print("weight check")

  #construct ligand_target_matrix
  ligand_target_matrix_EGF_ATAC <- construct_ligand_target_matrix(
  weighted_networks = weighted_networks_EGF_ATAC, 
  ligands = list("EGF"), algorithm = "PPR",
  damping_factor = hyperparameter_list %>% 
    filter(parameter == "damping_factor") %>% 
    pull(avg_weight),
  ltf_cutoff = hyperparameter_list %>% 
    filter(parameter == "ltf_cutoff") %>% 
    pull(avg_weight))
  
  ligand_target_matrix_HGF_ATAC <- construct_ligand_target_matrix(
  weighted_networks = weighted_networks_HGF_ATAC, 
  ligands = list("HGF"), algorithm = "PPR",
  damping_factor = hyperparameter_list %>% 
    filter(parameter == "damping_factor") %>% 
    pull(avg_weight),
  ltf_cutoff = hyperparameter_list %>% 
    filter(parameter == "ltf_cutoff") %>% 
    pull(avg_weight))
  
  ligand_target_matrix_OSM_ATAC <- construct_ligand_target_matrix(
  weighted_networks = weighted_networks_OSM_ATAC, 
  ligands = list("OSM"), algorithm = "PPR",
  damping_factor = hyperparameter_list %>% 
    filter(parameter == "damping_factor") %>% 
    pull(avg_weight),
  ltf_cutoff = hyperparameter_list %>% 
    filter(parameter == "ltf_cutoff") %>% 
    pull(avg_weight))
  
  all_weighted_networks_EGF_ATAC$per <- weighted_networks_EGF_ATAC
  all_weighted_networks_HGF_ATAC$per <- weighted_networks_HGF_ATAC
  all_weighted_networks_OSM_ATAC$per <- weighted_networks_OSM_ATAC
  
  all_ligand_target_matrix_EGF_ATAC$per <- ligand_target_matrix_EGF_ATAC
  all_ligand_target_matrix_HGF_ATAC$per <- ligand_target_matrix_HGF_ATAC
  all_ligand_target_matrix_OSM_ATAC$per <- ligand_target_matrix_OSM_ATAC
  
  print("ligand_target_matrix check")

  all_grn_OSM_ATAC$per <- as.data.frame(grn_OSM_ATAC)
  all_grn_EGF_ATAC$per <- as.data.frame(grn_EGF_ATAC)
  all_grn_HGF_ATAC$per <- as.data.frame(grn_HGF_ATAC)
  names_per <- append(names_per, toString(per))
  names(all_grn_OSM_ATAC) <- names_per
  names(all_grn_EGF_ATAC) <- names_per
  names(all_grn_HGF_ATAC) <- names_per
  names(all_ligand_target_matrix_EGF_ATAC) <- names_per
  names(all_ligand_target_matrix_HGF_ATAC) <- names_per
  names(all_ligand_target_matrix_OSM_ATAC) <- names_per
  names(all_weighted_networks_EGF_ATAC) <- names_per
  names(all_weighted_networks_HGF_ATAC) <- names_per
  names(all_weighted_networks_OSM_ATAC) <- names_per
  
}


```
## Evaluation NicheNet: Mock analysis
# Transcriptional response prediction evaluation
```{r model evaluation with per with ATAC+RNA, echo=TRUE, error=TRUE, warning=TRUE}
#one ligand (EGF, HGF, OSM)

ligands <- c("EGF", "OSM", "HGF")
all_performances_list <- list()
for(i in ligands){
expression_settings_validation_DE_one <- list(designs[[1]]$expression_settings_validation_DE[[i]])

names_one <- c(i)
names(expression_settings_validation_DE_one) <- names_one

all_performances <- c()
names_per <- c()
for(per in c("0", "0.25", "0.5", "0.75", "1")){
  ligand_target_matrix <- get(paste0("all_ligand_target_matrix_", i))[[per]]
  if(length(expression_settings_validation_DE_one) > 0){
    # Step 1: convert expression datasets to the required format to perform target gene prediction
    settings_one = expression_settings_validation_DE_one %>% lapply(convert_expression_settings_evaluation)
    settings_one = settings_one %>% discard(~length(.$from) > 1)
    designs[[1]]$settings_one <- settings_one
    #print(names(which(settings$response)))

    # Step 2: calculate the target gene prediction performances
    performances_one = settings_one %>% lapply(evaluate_target_prediction, ligand_target_matrix) %>% bind_rows()
    designs[[1]]$performances_one <- as.data.frame(performances_one)

    #writexl::write_xlsx(performances_one, path = paste0("NicheNet_output/performances_one_ANANSE_48_", per, ".xlsx")

    # Visualize some classification evaluation metrics showing the target gene prediction performance
    performances_one = performances_one %>% dplyr::select(-aupr, -auc_iregulon,-pearson_log_pval,-spearman_log_pval , -sensitivity_roc, -specificity_roc) %>% gather(key = scorename, value = scorevalue, auroc:spearman)
    scorelabels = c(auroc = "AUROC", aupr_corrected = "AUPR (corrected)", auc_iregulon_corrected = "AUC-iRegulon (corrected)",
                    pearson = "Pearson correlation", spearman = "Spearman's rank correlation",
                    mean_rank_GST_log_pval = "Mean-rank gene-set enrichment")
    scorerandom = c(auroc = 0.5, aupr_corrected = 0, auc_iregulon_corrected = 0, pearson = 0, spearman = 0, mean_rank_GST_log_pval = 0) %>% data.frame(scorevalue = .) %>% rownames_to_column("scorename")
    names_per <- append(names_per, per)

    all_performances$per <- performances_one
    names(all_performances) <- names_per

    print(performances_one %>%
            mutate(model = "NicheNet") %>%
            ggplot() +
            geom_violin(aes(model, scorevalue, group = model, fill = model)) +
            geom_boxplot(aes(model, scorevalue, group = model), width = 0.05) +
            scale_y_continuous("Score target prediction") +
            facet_wrap(~scorename, scales = "free", labeller = as_labeller(scorelabels)) +
            geom_hline(aes(yintercept = scorevalue), data = scorerandom, linetype = 2, color = "red") +
            theme_bw() +
            ggtitle(paste("48h with ANANSE data only one ligand", per))
    )
  }
  all_performances_list[[i]] <- all_performances
}
}
```

```{r model evaluation with per with ATAC, echo=TRUE, error=TRUE, warning=TRUE}
#one ligand (EGF, HGF, OSM)

ligands <- c("EGF", "OSM", "HGF")
all_performances_list_ATAC <- list()
for(i in ligands){
expression_settings_validation_DE_one <- list(designs[[1]]$expression_settings_validation_DE[[i]])

names_one <- c(i)
names(expression_settings_validation_DE_one) <- names_one

all_performances_ATAC <- c()
names_per <- c()
for(per in c("0", "0.25", "0.5", "0.75", "1")){
  ligand_target_matrix_ATAC <- get(paste0("all_ligand_target_matrix_", i, "_ATAC"))[[per]]
  if(length(expression_settings_validation_DE_one) > 0){
    # Step 1: convert expression datasets to the required format to perform target gene prediction
    settings_one = expression_settings_validation_DE_one %>% lapply(convert_expression_settings_evaluation)
    settings_one = settings_one %>% discard(~length(.$from) > 1)
    designs[[1]]$settings_one <- settings_one
    #print(names(which(settings$response)))

    # Step 2: calculate the target gene prediction performances
    performances_one_ATAC = settings_one %>% lapply(evaluate_target_prediction, ligand_target_matrix_ATAC) %>% bind_rows()
    designs[[1]]$performances_one_ATAC <- as.data.frame(performances_one_ATAC)

    #writexl::write_xlsx(performances_one, path = paste0("NicheNet_output/performances_one_ANANSE_48_", per, ".xlsx")

    # Visualize some classification evaluation metrics showing the target gene prediction performance
    performances_one_ATAC = performances_one_ATAC %>% dplyr::select(-aupr, -auc_iregulon, -pearson_log_pval,-spearman_log_pval , -sensitivity_roc, -specificity_roc) %>% gather(key = scorename, value = scorevalue, auroc:spearman)
    scorelabels = c(auroc = "AUROC", aupr_corrected = "AUPR (corrected)", auc_iregulon_corrected = "AUC-iRegulon (corrected)",
                    pearson = "Pearson correlation", spearman = "Spearman's rank correlation",
                    mean_rank_GST_log_pval = "Mean-rank gene-set enrichment")
    scorerandom = c(auroc = 0.5, aupr_corrected = 0, auc_iregulon_corrected = 0, pearson = 0, spearman = 0, mean_rank_GST_log_pval = 0) %>% data.frame(scorevalue = .) %>% rownames_to_column("scorename")
    names_per <- append(names_per, per)

    all_performances_ATAC$per <- performances_one_ATAC
    names(all_performances_ATAC) <- names_per

    print(performances_one_ATAC %>%
            mutate(model = "NicheNet") %>%
            ggplot() +
            geom_violin(aes(model, scorevalue, group = model, fill = model)) +
            geom_boxplot(aes(model, scorevalue, group = model), width = 0.05) +
            scale_y_continuous("Score target prediction") +
            facet_wrap(~scorename, scales = "free", labeller = as_labeller(scorelabels)) +
            geom_hline(aes(yintercept = scorevalue), data = scorerandom, linetype = 2, color = "red") +
            theme_bw() +
            ggtitle(paste("48h with ANANSE data only one ligand", per))
    )
  }
  all_performances_list_ATAC[[i]] <- all_performances_ATAC
}
}
```


#Statsplots with 0 to 100% DE genes against NicheNet with the 3 ligands: Mock analysis 

```{r statsplots performances mock analysis, echo=TRUE, error=TRUE, fig.width=10, fig.height = 10} 
performances_NicheNet["condition"] <- "NicheNet"  

all_perf_0 <- rbind(all_performances_list$EGF[["0"]], all_performances_list$HGF[["0"]], all_performances_list$OSM[["0"]])
all_perf_0.25 <- rbind(all_performances_list$EGF[["0.25"]], all_performances_list$HGF[["0.25"]], all_performances_list$OSM[["0.25"]])
all_perf_0.5 <- rbind(all_performances_list$EGF[["0.5"]], all_performances_list$HGF[["0.5"]], all_performances_list$OSM[["0.5"]])
all_perf_0.75 <- rbind(all_performances_list$EGF[["0.75"]], all_performances_list$HGF[["0.75"]], all_performances_list$OSM[["0.75"]])
all_perf_1 <- rbind(all_performances_list$EGF[["1"]], all_performances_list$HGF[["1"]], all_performances_list$OSM[["1"]])

all_perf_0_ATAC <- rbind(all_performances_list_ATAC$EGF[["0"]], all_performances_list_ATAC$HGF[["0"]], all_performances_list_ATAC$OSM[["0"]])
all_perf_0.25_ATAC <- rbind(all_performances_list_ATAC$EGF[["0.25"]], all_performances_list_ATAC$HGF[["0.25"]], all_performances_list_ATAC$OSM[["0.25"]])
all_perf_0.5_ATAC <- rbind(all_performances_list_ATAC$EGF[["0.5"]], all_performances_list_ATAC$HGF[["0.5"]], all_performances_list_ATAC$OSM[["0.5"]])
all_perf_0.75_ATAC <- rbind(all_performances_list_ATAC$EGF[["0.75"]], all_performances_list_ATAC$HGF[["0.75"]], all_performances_list_ATAC$OSM[["0.75"]])
all_perf_1_ATAC <- rbind(all_performances_list_ATAC$EGF[["1"]], all_performances_list_ATAC$HGF[["1"]], all_performances_list_ATAC$OSM[["1"]])

all_perf_0["condition"] <- "ATAC + RNA"  
all_perf_0.25["condition"] <- "ATAC + RNA" 
all_perf_0.5["condition"] <- "ATAC + RNA" 
all_perf_0.75["condition"] <- "ATAC + RNA" 
all_perf_1["condition"] <- "ATAC + RNA" 

all_perf_0_ATAC["condition"] <- "ATAC"  
all_perf_0.25_ATAC["condition"] <- "ATAC" 
all_perf_0.5_ATAC["condition"] <- "ATAC" 
all_perf_0.75_ATAC["condition"] <- "ATAC" 
all_perf_1_ATAC["condition"] <- "ATAC" 

perf_0 <- performances_NicheNet %>% rbind(all_perf_0)  %>% rbind(all_perf_0_ATAC)
perf_0.25 <- performances_NicheNet %>% rbind(all_perf_0.25) %>% rbind(all_perf_0.25_ATAC)
perf_0.5 <- performances_NicheNet %>% rbind(all_perf_0.5) %>% rbind(all_perf_0.5_ATAC)
perf_0.75 <- performances_NicheNet %>% rbind(all_perf_0.75) %>% rbind(all_perf_0.75_ATAC)
perf_1 <- performances_NicheNet %>% rbind(all_perf_1) %>% rbind(all_perf_1_ATAC)


perf_0$condition <- factor(perf_0$condition, levels = c("ATAC", "NicheNet", "ATAC + RNA"))
perf_0.25$condition <- factor(perf_0.25$condition, levels = c("ATAC", "NicheNet", "ATAC + RNA"))
perf_0.5$condition <- factor(perf_0.5$condition, levels = c("ATAC", "NicheNet", "ATAC + RNA"))
perf_0.75$condition <- factor(perf_0.75$condition, levels = c("ATAC", "NicheNet", "ATAC + RNA"))
perf_1$condition <- factor(perf_1$condition, levels = c("ATAC", "NicheNet", "ATAC + RNA"))

#auroc
au_0 <- perf_0[which(perf_0$scorename == "auroc"),] 
au_0.25 <- perf_0.25[which(perf_0.25$scorename == "auroc"),] 
au_0.5 <- perf_0.5[which(perf_0.5$scorename == "auroc"),] 
au_0.75 <- perf_0.75[which(perf_0.75$scorename == "auroc"),] 
au_1 <- perf_1[which(perf_1$scorename == "auroc"),] 

# Make AUC plots 
au_plot_0 <- ggwithinstats( 
   data = au_0, 
 x    = condition, 
 y    = scorevalue, 
   type = "np" 
) +
   ggtitle ("AUROC 0% DE genes, 24h") +
  coord_cartesian(ylim = c(0.65, 0.75)) +
  geom_text(aes(label = au_0$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

au_plot_0.25 <- ggwithinstats( 
   data = au_0.25, 
 x    = condition, 
 y    = scorevalue, 
   type = "np" 
) +
   ggtitle ("AUROC 25% DE genes, 24h") +
  coord_cartesian(ylim = c(0.65, 0.75)) +
  geom_text(aes(label = au_0.25$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

au_plot_0.5 <- ggwithinstats( 
   data = au_0.5, 
 x    = condition, 
 y    = scorevalue, 
   type = "np" 
) +
   ggtitle ("AUROC 50% DE genes, 24h") +
  coord_cartesian(ylim = c(0.65, 0.75)) +
  geom_text(aes(label = au_0.5$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

au_plot_0.75 <- ggwithinstats( 
   data = au_0.75, 
 x    = condition, 
 y    = scorevalue, 
   type = "np" 
) +
   ggtitle ("AUROC 75% DE genes, 24h") +
  coord_cartesian(ylim = c(0.65, 0.75)) +
  geom_text(aes(label = au_0.75$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

au_plot_1 <- ggwithinstats( 
   data = au_1, 
 x    = condition, 
 y    = scorevalue, 
   type = "np" 
) +
   ggtitle ("AUROC 100% DE genes, 24h") +
  coord_cartesian(ylim = c(0.65, 0.75)) +
  geom_text(aes(label = au_1$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

# Combine plots and save them
au_plot <- plot_grid(au_plot_0, au_plot_0.25, au_plot_0.5, au_plot_0.75, au_plot_1)
print(au_plot)
ggsave("Mock_au_24_ATAC_RNA.png", au_plot, height = 10, width = 15)

#aupr
aupr_0 <- perf_0[which(perf_0$scorename == "aupr_corrected"),] 
aupr_0.25 <- perf_0.25[which(perf_0.25$scorename == "aupr_corrected"),] 
aupr_0.5 <- perf_0.5[which(perf_0.5$scorename == "aupr_corrected"),] 
aupr_0.75 <- perf_0.75[which(perf_0.75$scorename == "aupr_corrected"),] 
aupr_1 <- perf_1[which(perf_1$scorename == "aupr_corrected"),] 

# Make aupr plots 
aupr_plot_0 <- ggwithinstats( 
   data = aupr_0, 
 x    = condition, 
 y    = scorevalue, 
   type = "np" 
) +
   ggtitle ("aupr 0% DE genes, 24h") +
  coord_cartesian(ylim = c(0, 0.2)) +
  geom_text(aes(label = aupr_0$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

aupr_plot_0.25 <- ggwithinstats( 
   data = aupr_0.25, 
 x    = condition, 
 y    = scorevalue, 
   type = "np" 
) +
   ggtitle ("aupr 25% DE genes, 24h") +
  coord_cartesian(ylim = c(0, 0.2)) +
  geom_text(aes(label = aupr_0.25$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

aupr_plot_0.5 <- ggwithinstats( 
   data = aupr_0.5, 
 x    = condition, 
 y    = scorevalue, 
   type = "np" 
) +
   ggtitle ("aupr 50% DE genes, 24h") +
  coord_cartesian(ylim = c(0, 0.2)) +
  geom_text(aes(label = aupr_0.5$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

aupr_plot_0.75 <- ggwithinstats( 
   data = aupr_0.75, 
 x    = condition, 
 y    = scorevalue, 
   type = "np" 
) +
   ggtitle ("aupr 75% DE genes, 24h") +
  coord_cartesian(ylim = c(0, 0.2)) +
  geom_text(aes(label = aupr_0.75$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

aupr_plot_1 <- ggwithinstats( 
   data = aupr_1, 
 x    = condition, 
 y    = scorevalue, 
   type = "np" 
) +
   ggtitle ("Aupr 100% DE genes, 24h") +
  coord_cartesian(ylim = c(0, 0.2)) +
  geom_text(aes(label = aupr_1$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

# Combine the plots
aupr_plot <- plot_grid(aupr_plot_0, aupr_plot_0.25, aupr_plot_0.5, aupr_plot_0.75, aupr_plot_1)
print(aupr_plot)
ggsave("Mock_aupr_24_ATAC_RNA.png", aupr_plot, height = 10, width = 15)

# Spearman
sp_0 <- perf_0[which(perf_0$scorename == "spearman"),] 
sp_0.25 <- perf_0.25[which(perf_0.25$scorename == "spearman"),] 
sp_0.5 <- perf_0.5[which(perf_0.5$scorename == "spearman"),] 
sp_0.75 <- perf_0.75[which(perf_0.75$scorename == "spearman"),] 
sp_1 <- perf_1[which(perf_1$scorename == "spearman"),] 

# Make spearman plots 
sp_plot_0 <- ggwithinstats( 
   data = sp_0, 
 x    = condition, 
 y    = scorevalue, 
   type = "np" 
) +
   ggtitle ("Spearman 0% DE genes, 24h") +
  coord_cartesian(ylim = c(0, 0.3)) +
  geom_text(aes(label = sp_0$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

sp_plot_0.25 <- ggwithinstats( 
   data = sp_0.25, 
 x    = condition, 
 y    = scorevalue, 
   type = "np" 
) +
   ggtitle ("Spearman 25% DE genes, 24h") +
  coord_cartesian(ylim = c(0, 0.3)) +
  geom_text(aes(label = sp_0.25$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

sp_plot_0.5 <- ggwithinstats( 
   data = sp_0.5, 
 x    = condition, 
 y    = scorevalue, 
   type = "np" 
) +
   ggtitle ("Spearman 50% DE genes, 24h") +
  coord_cartesian(ylim = c(0, 0.3)) +
  geom_text(aes(label = sp_0.5$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

sp_plot_0.75 <- ggwithinstats( 
   data = sp_0.75, 
 x    = condition, 
 y    = scorevalue, 
   type = "np" 
) +
   ggtitle ("Spearman 75% DE genes, 24h") +
  coord_cartesian(ylim = c(0, 0.3)) +
  geom_text(aes(label = sp_0.75$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

sp_plot_1 <- ggwithinstats( 
   data = sp_1, 
 x    = condition, 
 y    = scorevalue, 
   type = "np" 
) +
   ggtitle ("Spearman 100% DE genes, 24h") +
  coord_cartesian(ylim = c(0, 0.3)) +
  geom_text(aes(label = sp_1$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

# Combine the plots
sp_plot <- plot_grid(sp_plot_0, sp_plot_0.25, sp_plot_0.5, sp_plot_0.75, sp_plot_1)
print(sp_plot)
ggsave("Mock_sp_24_ATAC_RNA.png", sp_plot, height = 10, width = 15)

# Pearson
pe_0 <- perf_0[which(perf_0$scorename == "pearson"),] 
pe_0.25 <- perf_0.25[which(perf_0.25$scorename == "pearson"),] 
pe_0.5 <- perf_0.5[which(perf_0.5$scorename == "pearson"),] 
pe_0.75 <- perf_0.75[which(perf_0.75$scorename == "pearson"),] 
pe_1 <- perf_1[which(perf_1$scorename == "pearson"),] 

# Make spearman plots 
pe_plot_0 <- ggwithinstats( 
   data = pe_0, 
 x    = condition, 
 y    = scorevalue, 
   type = "np" 
) +
   ggtitle ("Pearson 0% DE genes, 24h") +
  coord_cartesian(ylim = c(0, 0.2)) +
  geom_text(aes(label = pe_0$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

pe_plot_0.25 <- ggwithinstats( 
   data = pe_0.25, 
 x    = condition, 
 y    = scorevalue, 
   type = "np" 
) +
   ggtitle ("Pearson 25% DE genes, 24h") +
  coord_cartesian(ylim = c(0, 0.2)) +
  geom_text(aes(label = pe_0.25$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

pe_plot_0.5 <- ggwithinstats( 
   data = pe_0.5, 
 x    = condition, 
 y    = scorevalue, 
   type = "np" 
) +
   ggtitle ("Pearson 50% DE genes, 24h") +
  coord_cartesian(ylim = c(0, 0.2)) +
  geom_text(aes(label = pe_0.5$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

pe_plot_0.75 <- ggwithinstats( 
   data = pe_0.75, 
 x    = condition, 
 y    = scorevalue, 
   type = "np" 
) +
   ggtitle ("Pearson 75% DE genes, 24h") +
  coord_cartesian(ylim = c(0, 0.2)) +
  geom_text(aes(label = pe_0.75$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

pe_plot_1 <- ggwithinstats( 
   data = pe_1, 
 x    = condition, 
 y    = scorevalue, 
   type = "np" 
) +
   ggtitle ("Pearson 100% DE genes, 24h") +
  coord_cartesian(ylim = c(0, 0.2)) +
  geom_text(aes(label = pe_1$setting), size = 3,  hjust = 2.5, size = 20, fontface = "bold") +
  theme(plot.title = element_text(hjust = 0.5))

# Combine the plots
pe_plot <- plot_grid(pe_plot_0, pe_plot_0.25, pe_plot_0.5, pe_plot_0.75, pe_plot_1)
print(pe_plot)
ggsave("Mock_pe_24_ATAC_RNA.png", pe_plot, height = 10, width = 15)
``` 


