---
title: "Bioxrvir_with_ANANSE_24"
author: "Klara Dewitte"
date: "2022-11-28"
output: html_document
---
##Load packages, get RNA-seq data and get DE genes
Load the packages.
```{r libraries}
library(limma) 
library(edgeR)
library(stringr)
library(tidyverse)
library(dplyr)
library(nichenetr)
library(biomaRt)
library(data.table)
library(stringr)
library(synapser) 
library(knitr)
library(ggstatsplot)
library(cowplot)
```

Get the RNAsesq data from Synapsr.
```{r loop, echo=TRUE}
# login of synapsr
synLogin('KlaraDewitte','Thesis2022') 
 
# Obtain a pointer and download the RNA-seq level 1 data
syn18508389 <- synGet(entity = 'syn18508389') 
filepath_RNA <- syn18508389$path
rawdata <- read.csv(filepath_RNA, row.names = 1)

# Obtain pointer and download metadata
ssyn15574180 <- synGet(entity = 'syn15574180') 
filepath_metadata <- ssyn15574180$path
metadata <- read.csv(filepath_metadata)

# Obtain subset of metadata and RNA-seq data for 24h and 48h
values <- stringr::str_split(metadata$specimenName, "_") %>% 
                  do.call(rbind, .) 

id_24 <- metadata[which(values[,2] == 24 & metadata$RNAseq_QCpass == TRUE) ,]$specimenID 
id_48 <- metadata[which(values[,2] == 48 & metadata$RNAseq_QCpass == TRUE),]$specimenID

rawdata_subset_24 <- rawdata[,which(colnames(rawdata) %in% id_24)]
rawdata_subset_48 <- rawdata[,which(colnames(rawdata) %in% id_48)]

metadata_subset_24 <- metadata[which(metadata$specimenID %in% id_24),]
metadata_subset_48 <- metadata[which(metadata$specimenID %in% id_48),]

# Get unique values of each cytokin, timepoint and replicate
unique_values <- values %>%
                  apply(2, unique) %>%
                  setNames(c("cytokine", "timepoint", "collection", "replicate"))

ensembl <- useEnsembl(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")

# Get treatment groups for 24h
Treat <- sapply(str_split(metadata_subset_24$specimenName, "_"), function(k) k[1])
Treat <- Treat %>% factor %>% relevel(ref = "PBS")

# Create design matrix
design <- model.matrix(~ Treat)
rownames(design) <- metadata_subset_24$specimenName
    
# Preprocess counts, estimate dispersion
y <- DGEList(counts = rawdata_subset_24, group = Treat)
y <- calcNormFactors(y)
y <- estimateDisp(y, design, robust = TRUE)
    
# Fit the model
fit <- glmQLFit(y, design)
    
# Perform test for each cytokine against PBS
res <- data.table()
for (treat in colnames(design)[-1]){
  qlf <- glmQLFTest(fit, coef = treat)
  DE_genes_table <- data.frame(topTags(qlf, sort.by = "PValue", n = Inf),  coef = substring(treat, 6))
  DE_genes_table$Gene <- rownames(DE_genes_table)
  res <- rbind(res, DE_genes_table)
}
   
# Add gene names
IDtoSymbol_matrix <- getBM(attributes = c("ensembl_gene_id", "external_gene_name"), filters = "ensembl_gene_id", values = unique(res$Gene), mart = ensembl)
top_table <- merge(res, IDtoSymbol_matrix, by.x = 'Gene', by.y = 'ensembl_gene_id', all.x = TRUE)
    
# Omit all the genes with NA or an empty name
top_na <- na.omit(top_table) 
top_na <- top_na[!(top_na$external_gene_name==""),]
    
# Select DE genes and make expression_settings_validation for evaluation NicheNet
total <- c()
total_names_DE <- c()
DEgenes <- list
expression_settings_validation_DE <- list()
DE_info <- list()
designs <- list()
table1 <- list()
designs <- list()

for(treatx in unique_values$cytokine){
  if (treatx != "PBS" & treatx != "ctrl"){
    top_na_subset <- top_na[top_na$coef == treatx,]
    DE <- which(top_na_subset$FDR < 0.05 & abs(top_na_subset$logFC) > 2)
    DEgenes <- top_na_subset[DE,]
        
    table2 <- list()
    table2$DEgenes <- DEgenes
    DE_info <- append(DE_info, list(DEgenes))
    
    treatx <- gsub("TGFB", "TGFB1", treatx)
    treatx <- gsub("IFNg", "IFNG", treatx)
        
    print(treatx)
    print(paste0("number of DE genes ", length(DE)))
    
    if (length(DE) > 0){
      total_names_DE <- append(total_names_DE, treatx)
      subsets <- subset(top_na_subset, select = c(logFC, FDR, external_gene_name))
      names(subsets) <- c('lfc', 'qval', 'gene')
      rownames(subsets) <- 1:nrow(subsets)
      table3 <- list()
      table3$diffexp <- as.data.frame(subsets)
      table3$from <- c(treatx)
      table3$name <- c(treatx)
          
      expression_settings_validation_DE <- append(expression_settings_validation_DE, list(table3))
    }
    total <- (append(total, treatx))
    names(expression_settings_validation_DE) <- total_names_DE
    table1$expression_settings_validation_DE <- expression_settings_validation_DE 
        
   }
}
names(DE_info) <- total
table1$DE_info <- DE_info
designs <- append(designs, list(table1))
```

# Evaluation without ATAC-seq
## NicheNet without any modifications
### Ligand_target_matrix included in NicheNet
```{r NicehNet ligand_target_matrix, echo=TRUE}
file_ligand <- "Datasets_vignettes/ligand_target_matrix_nsga2r_final.rds"
ligand_target_matrix_NicheNet <- readRDS(file = file_ligand)
```

### Transcriptional response prediction evaluation without ATAC-seq, with NicheNet ligand_target_matrix

```{r model evaluation without ATAC-seq NicheNet matrix, echo=TRUE, error=TRUE, warning=TRUE}
#one ligand (EGF, HGF, OSM)
expression_settings_validation_DE_one <- list(designs[[1]]$expression_settings_validation_DE$EGF,
                                designs[[1]]$expression_settings_validation_DE$HGF,
                                designs[[1]]$expression_settings_validation_DE$OSM)

names_one <- c("EGF", "HGF", "OSM")
names(expression_settings_validation_DE_one) <- names_one

if(length(expression_settings_validation_DE_one) > 0){
  # Step 1: convert expression datasets to the required format to perform target gene prediction
  settings_one = expression_settings_validation_DE_one %>% lapply(convert_expression_settings_evaluation)
  settings_one = settings_one %>% discard(~length(.$from) > 1)
  designs[[1]]$settings_one <- settings_one
  #print(names(which(settings$response)))
  
  # Step 2: calculate the target gene prediction performances
  performances_one = settings_one %>% lapply(evaluate_target_prediction, ligand_target_matrix_NicheNet) %>% bind_rows() 
  designs[[1]]$performances_one <- as.data.frame(performances_one)
    
  # Visualize some classification evaluation metrics showing the target gene prediction performance
  performances_NicheNet = performances_one %>% dplyr::select(-aupr, -auc_iregulon, -pearson_log_pval, -spearman_log_pval,
                                                             -sensitivity_roc, -specificity_roc) %>% 
    gather(key = scorename, value = scorevalue, auroc:spearman)
  scorelabels = c(auroc = "AUROC", aupr_corrected = "AUPR (corrected)", auc_iregulon_corrected = "AUC-iRegulon (corrected)", 
                  pearson = "Pearson correlation", spearman = "Spearman's rank correlation", 
                  mean_rank_GST_log_pval = "Mean-rank gene-set enrichment")
  scorerandom = c(auroc = 0.5, aupr_corrected = 0, auc_iregulon_corrected = 0, pearson = 0, spearman = 0, 
                  mean_rank_GST_log_pval = 0) %>% data.frame(scorevalue = .) %>% rownames_to_column("scorename")
    
  print(performances_NicheNet %>%
          mutate(model = "NicheNet") %>%
          ggplot() +
          geom_violin(aes(model, scorevalue, group = model, fill = model)) +
          geom_boxplot(aes(model, scorevalue, group = model), width = 0.05) +
          scale_y_continuous("Score target prediction") +
          facet_wrap(~scorename, scales = "free", labeller = as_labeller(scorelabels)) +
          geom_hline(aes(yintercept = scorevalue), data = scorerandom, linetype = 2, color = "red") +
          theme_bw()
  )
}
```
### ligand activity prediction evaluation without ATAC-seq, wiht Nichnet ligand_target_matrix

```{r ligand activity prediction evaluation without ATAC-seq NicheNet, echo=TRUE, error=TRUE}
#one ligand
designs[[1]]$settings_one <- settings_one
names(settings_one) <- names_one
    
if(length(settings_one) > 0){
  # convert expression datasets to correct format for ligand activity prediction
  all_ligands = settings_one %>% extract_ligands_from_settings(combination = FALSE) %>% unlist()
  settings_ligand_prediction_one = settings_one %>% convert_settings_ligand_prediction(all_ligands = all_ligands, validation = TRUE)
    
   # infer ligand importances: for all ligands of interest, we assess how well a ligand explains the differential expression in a specific dataset (and we do this for all datasets).
   ligand_importances_one = settings_ligand_prediction_one %>% lapply(get_single_ligand_importances, ligand_target_matrix_NicheNet) %>% bind_rows()
   ligand_importances_one <- do.call(data.frame, lapply(ligand_importances_one, function(value) replace(value, is.infinite(value), 100000)))
    
    # Look at predictive performance of single/individual importance measures to predict ligand activity: of all ligands tested, the ligand that is truly active in a dataset should get the highest activity score (i.e. best target gene prediction performance)
   evaluation_ligand_prediction_NicheNet = ligand_importances_one$setting %>% unique() %>% lapply(function(x){x}) %>% 
    lapply(wrapper_evaluate_single_importances_ligand_prediction, ligand_importances_one) %>%
    bind_rows() %>% inner_join(ligand_importances_one %>% distinct(setting, ligand))
   
    
  # Visualize some classification evaluation metrics showing the ligand activity prediction performance
  evaluation_ligand_prediction_NicheNet = evaluation_ligand_prediction_NicheNet %>% dplyr::select(-aupr, -sensitivity_roc, -specificity_roc,
                                                                                        -pearson, -spearman,
                                                                                        -mean_rank_GST_log_pval) %>% 
    gather(key = scorename, value = scorevalue, auroc:aupr_corrected)
  scorelabels = c(auroc = "AUROC", aupr_corrected = "AUPR (corrected)")
  scorerandom = c(auroc = 0.5, aupr_corrected = 0) %>% data.frame(scorevalue = .) %>% rownames_to_column("scorename")
    
  print(evaluation_ligand_prediction_NicheNet %>%
     filter(importance_measure %in% c("auroc", "aupr_corrected", "mean_rank_GST_log_pval", "auc_iregulon_corrected", "pearson",
                                      "spearman")) %>%
    ggplot() +
    geom_violin(aes(importance_measure, scorevalue, group = importance_measure, fill = importance_measure)) +
    geom_boxplot(aes(importance_measure, scorevalue, group = importance_measure), width = 0.1) +
    scale_y_continuous("Evaluation ligand activity prediction") +
    scale_x_discrete("Ligand activity measure") +
    facet_wrap(~scorename, scales = "free", labeller = as_labeller(scorelabels)) +
    geom_hline(aes(yintercept = scorevalue), data = scorerandom, linetype = 2, color = "red") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90))) +
    ggtitle("24h with ANANSE data only one ligand")
  }

```

## NicheNet with construction of own ligand_target_matrix, only containing our ligands
### New ligand-target matrix, with our ligands
```{r ligand-target-matrix construction new matrix, echo=TRUE}
# NicheNet default networks
lr_network <- readRDS(url("https://zenodo.org/record/5884439/files/lr_network_human_21122021.rds"))
sig_network <- readRDS(url("https://zenodo.org/record/5884439/files/signaling_network_human_21122021.rds"))
gr_network <- readRDS(url("https://zenodo.org/record/5884439/files/gr_network_human_21122021.rds"))

# aggregate the individual data sources in a weighted manner to obtain a weighted integrated signaling network
weighted_networks_NicheNet_ligands = construct_weighted_networks(lr_network = lr_network, sig_network = sig_network, gr_network = gr_network, source_weights_df = source_weights_df)

ligands <- list("EGF", "HGF", "OSM")

ligand_target_matrix_NicheNet_ligands <- construct_ligand_target_matrix(weighted_networks = weighted_networks_NicheNet_ligands, ligands = ligands)
```

### Transcriptional response prediction evaluation without ATAC-seq, with new ligand_target_matrix

```{r model evaluation without ATAC-seq own ligand_target_matrix, echo=TRUE, error=TRUE, warning=TRUE}
#one ligand (EGF, HGF, OSM)
expression_settings_validation_DE_one <- list(designs[[1]]$expression_settings_validation_DE$EGF,
                                designs[[1]]$expression_settings_validation_DE$HGF,
                                designs[[1]]$expression_settings_validation_DE$OSM)

names_one <- c("EGF", "HGF", "OSM")
names(expression_settings_validation_DE_one) <- names_one

if(length(expression_settings_validation_DE_one) > 0){
  # Step 1: convert expression datasets to the required format to perform target gene prediction
  settings_one = expression_settings_validation_DE_one %>% lapply(convert_expression_settings_evaluation)
  settings_one = settings_one %>% discard(~length(.$from) > 1)
  designs[[1]]$settings_one <- settings_one
  #print(names(which(settings$response)))
  
  # Step 2: calculate the target gene prediction performances
  performances_one = settings_one %>% lapply(evaluate_target_prediction, ligand_target_matrix_NicheNet_ligands) %>% bind_rows() 
  designs[[1]]$performances_one <- as.data.frame(performances_one)
    
  # Visualize some classification evaluation metrics showing the target gene prediction performance
  performances_NicheNet_ligands = performances_one %>% dplyr::select(-aupr, -auc_iregulon, -pearson_log_pval, -spearman_log_pval,
                                                                     -sensitivity_roc, -specificity_roc) %>% 
    gather(key = scorename, value = scorevalue, auroc:spearman)
  scorelabels = c(auroc = "AUROC", aupr_corrected = "AUPR (corrected)", auc_iregulon_corrected = "AUC-iRegulon (corrected)", 
                  pearson = "Pearson correlation", spearman = "Spearman's rank correlation", 
                  mean_rank_GST_log_pval = "Mean-rank gene-set enrichment")
  scorerandom = c(auroc = 0.5, aupr_corrected = 0, auc_iregulon_corrected = 0, pearson = 0, spearman = 0, 
                  mean_rank_GST_log_pval = 0) %>% data.frame(scorevalue = .) %>% rownames_to_column("scorename")
    
  print(performances_NicheNet_ligands %>%
          mutate(model = "NicheNet") %>%
          ggplot() +
          geom_violin(aes(model, scorevalue, group = model, fill = model)) +
          geom_boxplot(aes(model, scorevalue, group = model), width = 0.05) +
          scale_y_continuous("Score target prediction") +
          facet_wrap(~scorename, scales = "free", labeller = as_labeller(scorelabels)) +
          geom_hline(aes(yintercept = scorevalue), data = scorerandom, linetype = 2, color = "red") +
          theme_bw()
  )
}

```

### ligand activity prediction evaluation without ATAC-seq, wiht new ligand_target_matrix

```{r ligand activity prediction evaluation without ATAC-seq new matrix, echo=TRUE, error=TRUE}
#one ligand
designs[[1]]$settings_one <- settings_one
names(settings_one) <- names_one
    
if(length(settings_one) > 0){
  # convert expression datasets to correct format for ligand activity prediction
  all_ligands = settings_one %>% extract_ligands_from_settings(combination = FALSE) %>% unlist()
  settings_ligand_prediction_one = settings_one %>% convert_settings_ligand_prediction(all_ligands = all_ligands, validation = TRUE)
    
   # infer ligand importances: for all ligands of interest, we assess how well a ligand explains the differential expression in a specific dataset (and we do this for all datasets).
   ligand_importances_NicheNet_ligands = settings_ligand_prediction_one %>% lapply(get_single_ligand_importances, ligand_target_matrix_NicheNet_ligands) %>% bind_rows()
   ligand_importances_NicheNet_ligands <- do.call(data.frame, lapply(ligand_importances_NicheNet_ligands, function(value) replace(value, is.infinite(value), 100000)))
    
    # Look at predictive performance of single/individual importance measures to predict ligand activity: of all ligands tested, the ligand that is truly active in a dataset should get the highest activity score (i.e. best target gene prediction performance)
   evaluation_ligand_prediction_NicheNet_ligands = ligand_importances_NicheNet_ligands$setting %>% unique() %>% lapply(function(x){x}) %>% 
    lapply(wrapper_evaluate_single_importances_ligand_prediction, ligand_importances_one) %>%
    bind_rows() %>% inner_join(ligand_importances_one %>% distinct(setting, ligand))
   
    
  # Visualize some classification evaluation metrics showing the ligand activity prediction performance
  evaluation_ligand_prediction_NicheNet_ligands = evaluation_ligand_prediction_NicheNet_ligands %>% 
                                                dplyr::select(-aupr, -sensitivity_roc, -specificity_roc,-pearson, -spearman,
                                                                                        -mean_rank_GST_log_pval) %>% 
    gather(key = scorename, value = scorevalue, auroc:aupr_corrected)
  scorelabels = c(auroc = "AUROC", aupr_corrected = "AUPR (corrected)")
  scorerandom = c(auroc = 0.5, aupr_corrected = 0) %>% data.frame(scorevalue = .) %>% rownames_to_column("scorename")
    
  print(evaluation_ligand_prediction_NicheNet_ligands %>%
     filter(importance_measure %in% c("auroc", "aupr_corrected", "mean_rank_GST_log_pval", "auc_iregulon_corrected", "pearson",
                                      "spearman")) %>%
    ggplot() +
    geom_violin(aes(importance_measure, scorevalue, group = importance_measure, fill = importance_measure)) +
    geom_boxplot(aes(importance_measure, scorevalue, group = importance_measure), width = 0.1) +
    scale_y_continuous("Evaluation ligand activity prediction") +
    scale_x_discrete("Ligand activity measure") +
    facet_wrap(~scorename, scales = "free", labeller = as_labeller(scorelabels)) +
    geom_hline(aes(yintercept = scorevalue), data = scorerandom, linetype = 2, color = "red") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90))) +
    ggtitle("24h with ANANSE data only one ligand")
  }

```

##Construction of ligand_target_matrix
In GRN.R are the weighted networks and individual ligand-target-matrices constructed.
```{r ligand-target-matrix construction, echo=TRUE}
# NicheNet default networks
# lr_network <- readRDS(url("https://zenodo.org/record/5884439/files/lr_network_human_21122021.rds"))
# sig_network <- readRDS(url("https://zenodo.org/record/5884439/files/signaling_network_human_21122021.rds"))
gr_network <- readRDS(url("https://zenodo.org/record/5884439/files/gr_network_human_21122021.rds"))

#import one ligand GRN networks
gr_network_LINCS_OSM <- read.delim("Ananse_networks/GRN/LINCS_MCF10A_OSM_24_NO_RNAseq_grn.txt", header = TRUE, sep = ";")
gr_network_LINCS_HGF <- read.delim("Ananse_networks/GRN/LINCS_MCF10A_HGF_24_NO_RNAseq_grn.txt", header = TRUE, sep = ";")
gr_network_LINCS_EGF <- read.delim("Ananse_networks/GRN/LINCS_MCF10A_EGF_24_NO_RNAseq_grn.txt", header = TRUE, sep = ";")

#add your new source to the weighted networks in NicheNet (weight??)
new_network_weights_df <- tibble(source = "LINCS_MCF10A", weight = 1)
new_source_weights_df <- optimized_source_weights_df %>% bind_rows(new_network_weights_df)

#add LINCS GRN to default GRN NicheNet
new_gr_network_total <- gr_network %>% bind_rows(gr_network_LINCS_OSM) %>% bind_rows(gr_network_LINCS_HGF) %>%
  bind_rows(gr_network_LINCS_EGF)
new_gr_network_total$source <- str_replace_all(new_gr_network_total$source, "LINCS_MCF10A_HGF_24_NO_RNAseq", "LINCS_MCF10A")
new_gr_network_total$source <- str_replace_all(new_gr_network_total$source, "LINCS_MCF10A_OSM_24_NO_RNAseq", "LINCS_MCF10A")
new_gr_network_total$source <- str_replace_all(new_gr_network_total$source, "LINCS_MCF10A_EGF_24_NO_RNAseq", "LINCS_MCF10A")


# aggregate the individual data sources in a weighted manner to obtain a weighted integrated signaling network
weighted_networks_total = construct_weighted_networks(lr_network = lr_network, sig_network = sig_network, gr_network = new_gr_network_total, source_weights_df = new_source_weights_df)

ligands <- list("EGF", "HGF", "OSM")

ligand_target_matrix_total <- construct_ligand_target_matrix(weighted_networks = weighted_networks_total, ligands = ligands)
```

## Evaluation NicheNet
# Transcriptional response prediction evaluation
```{r model evaluation, echo=TRUE, error=TRUE, warning=TRUE}
#one ligand (EGF, HGF, OSM)
expression_settings_validation_DE_one <- list(designs[[1]]$expression_settings_validation_DE$EGF,
                                designs[[1]]$expression_settings_validation_DE$HGF,
                                designs[[1]]$expression_settings_validation_DE$OSM)

names_one <- c("EGF", "HGF", "OSM")
names(expression_settings_validation_DE_one) <- names_one

if(length(expression_settings_validation_DE_one) > 0){
  # Step 1: convert expression datasets to the required format to perform target gene prediction
  settings_one = expression_settings_validation_DE_one %>% lapply(convert_expression_settings_evaluation)
  settings_one = settings_one %>% discard(~length(.$from) > 1)
  designs[[1]]$settings_one <- settings_one
  #print(names(which(settings$response)))
  
  # Step 2: calculate the target gene prediction performances
  performances_total = settings_one %>% lapply(evaluate_target_prediction, ligand_target_matrix_total) %>% bind_rows() 
  designs[[1]]$performances_total <- as.data.frame(performances_total)
  
    
  # Visualize some classification evaluation metrics showing the target gene prediction performance
  performances_total = performances_total %>% dplyr::select(-aupr, -auc_iregulon,-pearson_log_pval,-spearman_log_pval , -sensitivity_roc, -specificity_roc) %>% gather(key = scorename, value = scorevalue, auroc:spearman)
  scorelabels = c(auroc = "AUROC", aupr_corrected = "AUPR (corrected)", auc_iregulon_corrected = "AUC-iRegulon (corrected)", 
                  pearson = "Pearson correlation", spearman = "Spearman's rank correlation", 
                  mean_rank_GST_log_pval = "Mean-rank gene-set enrichment")
  scorerandom = c(auroc = 0.5, aupr_corrected = 0, auc_iregulon_corrected = 0, pearson = 0, spearman = 0, mean_rank_GST_log_pval = 0) %>% data.frame(scorevalue = .) %>% rownames_to_column("scorename")
    
  print(performances_total %>%
          mutate(model = "NicheNet") %>%
          ggplot() +
          geom_violin(aes(model, scorevalue, group = model, fill = model)) +
          geom_boxplot(aes(model, scorevalue, group = model), width = 0.05) +
          scale_y_continuous("Score target prediction") +
          facet_wrap(~scorename, scales = "free", labeller = as_labeller(scorelabels)) +
          geom_hline(aes(yintercept = scorevalue), data = scorerandom, linetype = 2, color = "red") +
          theme_bw() +
          ggtitle("24h with ANANSE data only one ligand")
  )
}
```

# ligand activity prediction evaluation

```{r ligand activity prediction evaluation, echo=TRUE, error=TRUE}
#one ligand
designs[[1]]$settings_one <- settings_one
names(settings_one) <- names_one
    
if(length(settings_one) > 0){
  # convert expression datasets to correct format for ligand activity prediction
  all_ligands = settings_one %>% extract_ligands_from_settings(combination = FALSE) %>% unlist()
  settings_ligand_prediction_one = settings_one %>% convert_settings_ligand_prediction(all_ligands = all_ligands, validation = TRUE)
    
   # infer ligand importances: for all ligands of interest, we assess how well a ligand explains the differential expression in a specific dataset (and we do this for all datasets).
   ligand_importances_one = settings_ligand_prediction_one %>% lapply(get_single_ligand_importances, ligand_target_matrix_total) %>% bind_rows()
   ligand_importances_one <- do.call(data.frame, lapply(ligand_importances_one, function(value) replace(value, is.infinite(value), 100000)))
    
    # Look at predictive performance of single/individual importance measures to predict ligand activity: of all ligands tested, the ligand that is truly active in a dataset should get the highest activity score (i.e. best target gene prediction performance)
   evaluation_ligand_prediction_one = ligand_importances_one$setting %>% unique() %>% lapply(function(x){x}) %>% 
    lapply(wrapper_evaluate_single_importances_ligand_prediction, ligand_importances_one) %>%
    bind_rows() %>% inner_join(ligand_importances_one %>% distinct(setting, ligand))
   
  writexl::write_xlsx(evaluation_ligand_prediction_one, path = "NicheNet_output/ligand_one_ANANSE_24.xlsx")
    
  # Visualize some classification evaluation metrics showing the ligand activity prediction performance
  evaluation_ligand_prediction_total = evaluation_ligand_prediction_one %>% dplyr::select(-aupr, -sensitivity_roc, -specificity_roc,
                                                                                        -pearson, -spearman,
                                                                                        -mean_rank_GST_log_pval) %>% 
    gather(key = scorename, value = scorevalue, auroc:aupr_corrected)
  scorelabels = c(auroc = "AUROC", aupr_corrected = "AUPR (corrected)")
  scorerandom = c(auroc = 0.5, aupr_corrected = 0) %>% data.frame(scorevalue = .) %>% rownames_to_column("scorename")
    
  print(evaluation_ligand_prediction_total %>%
     filter(importance_measure %in% c("auroc", "aupr_corrected", "mean_rank_GST_log_pval", "auc_iregulon_corrected", "pearson", "spearman")) %>%
    ggplot() +
    geom_violin(aes(importance_measure, scorevalue, group = importance_measure, fill = importance_measure)) +
    geom_boxplot(aes(importance_measure, scorevalue, group = importance_measure), width = 0.1) +
    scale_y_continuous("Evaluation ligand activity prediction") +
    scale_x_discrete("Ligand activity measure") +
    facet_wrap(~scorename, scales = "free", labeller = as_labeller(scorelabels)) +
    geom_hline(aes(yintercept = scorevalue), data = scorerandom, linetype = 2, color = "red") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90))) +
    ggtitle("24h with ANANSE data only one ligand")
  }

```

##Construction of ligand_target_matrix
In GRN.R are the weighted networks and individual ligand-target-matrices constructed.
```{r ligand-target-matrix construction percentages}
# The complete networks can be downloaded from Zenodo
lr_network <- readRDS(url("https://zenodo.org/record/5884439/files/lr_network_human_21122021.rds"))
sig_network <- readRDS(url("https://zenodo.org/record/5884439/files/signaling_network_human_21122021.rds"))
gr_network <- readRDS(url("https://zenodo.org/record/5884439/files/gr_network_human_21122021.rds"))

#import one ligand GRN networks
gr_network_LINCS_OSM <- read.delim("Ananse_networks/GRN/LINCS_MCF10A_OSM_24_NO_RNAseq_grn.txt", header = TRUE, sep = ";")
gr_network_LINCS_HGF <- read.delim("Ananse_networks/GRN/LINCS_MCF10A_HGF_24_NO_RNAseq_grn.txt", header = TRUE, sep = ";")
gr_network_LINCS_EGF <- read.delim("Ananse_networks/GRN/LINCS_MCF10A_EGF_24_NO_RNAseq_grn.txt", header = TRUE, sep = ";")

#only keep DE_genes in GRN
gr_network_LINCS_OSM_DE <- gr_network_LINCS_OSM[(gr_network_LINCS_OSM$to %in% table1$DE_info$OSM$external_gene_name),]
gr_network_LINCS_HGF_DE <- gr_network_LINCS_HGF[(gr_network_LINCS_HGF$to %in% table1$DE_info$HGF$external_gene_name),]
gr_network_LINCS_EGF_DE <- gr_network_LINCS_EGF[(gr_network_LINCS_EGF$to %in% table1$DE_info$EGF$external_gene_name),]

#GRN without DEgenes
gr_network_LINCS_OSM_not <- gr_network_LINCS_OSM[!(gr_network_LINCS_OSM$to %in% table1$DE_info$OSM$external_gene_name),]
gr_network_LINCS_HGF_not <- gr_network_LINCS_HGF[!(gr_network_LINCS_HGF$to %in% table1$DE_info$HGF$external_gene_name),]
gr_network_LINCS_EGF_not <- gr_network_LINCS_EGF[!(gr_network_LINCS_EGF$to %in% table1$DE_info$EGF$external_gene_name),]

#add your new source to the weighted networks in NicheNet (weight??)
new_network_weights_df <- tibble(source = "LINCS_MCF10A", weight = 1)
new_source_weights_df <- optimized_source_weights_df %>% bind_rows(new_network_weights_df)

#make new GRNs with 0,25,50,75 and 100% of DE genes
all_grn_OSM <- list()
all_grn_EGF <- list()
all_grn_HGF <- list()
names_per <- c()
all_ligand_target_matrix <- list()
all_new_grn <- c()
all_weighted_networks <- c()
for(per in c(0, 0.25, 0.5, 0.75, 1)){
  grn_OSM <- gr_network_LINCS_OSM_DE[sample(nrow(gr_network_LINCS_OSM_DE), round(nrow(gr_network_LINCS_OSM_DE)*per)), ] %>%
             rbind(gr_network_LINCS_OSM_not[sample(nrow(gr_network_LINCS_OSM_not),
                                                   round(nrow(gr_network_LINCS_OSM_DE)*(1-per))), ])
  grn_EGF <- gr_network_LINCS_EGF_DE[sample(nrow(gr_network_LINCS_EGF_DE), round(nrow(gr_network_LINCS_EGF_DE)*per)), ] %>%
             rbind(gr_network_LINCS_EGF_not[sample(nrow(gr_network_LINCS_EGF_not),
                                                   round(nrow(gr_network_LINCS_EGF_DE)*(1-per))), ])
  grn_HGF <- gr_network_LINCS_OSM_DE[sample(nrow(gr_network_LINCS_HGF_DE), round(nrow(gr_network_LINCS_HGF_DE)*per)), ] %>%
             rbind(gr_network_LINCS_HGF_not[sample(nrow(gr_network_LINCS_HGF_not),
                                                   round(nrow(gr_network_LINCS_HGF_DE)*(1-per))), ])
  new_gr_network <- gr_network %>% bind_rows(grn_HGF) %>% bind_rows(grn_EGF) %>% bind_rows(grn_OSM)
  new_gr_network$source <- str_replace_all(new_gr_network$source, "LINCS_MCF10A_HGF_24_NO_RNAseq", "LINCS_MCF10A")
  new_gr_network$source <- str_replace_all(new_gr_network$source, "LINCS_MCF10A_OSM_24_NO_RNAseq", "LINCS_MCF10A")
  new_gr_network$source <- str_replace_all(new_gr_network$source, "LINCS_MCF10A_EGF_24_NO_RNAseq", "LINCS_MCF10A")

  print(head(new_gr_network))
  print("grn check")
  
  weighted_networks = construct_weighted_networks(lr_network = lr_network, sig_network = sig_network, 
                                                  gr_network = new_gr_network, source_weights_df = new_source_weights_df)
  all_new_grn$per <- as.data.frame(new_gr_network)
  all_weighted_networks$per <- weighted_networks
  print(head(weighted_networks))
  print("weight check")
  
  #construct ligand_target_matrix
  ligands <- list("EGF", "HGF", "OSM")
  ligand_target_matrix <- construct_ligand_target_matrix(weighted_networks = weighted_networks, ligands = ligands)
  all_ligand_target_matrix$per <- ligand_target_matrix
  print(head(ligand_target_matrix))
  print("ligand_target_matrix check")

  all_grn_OSM$per <- as.data.frame(grn_OSM)
  all_grn_EGF$per <- as.data.frame(grn_EGF)
  all_grn_HGF$per <- as.data.frame(grn_HGF)
  names_per <- append(names_per, toString(per))
  names(all_grn_OSM) <- names_per
  names(all_grn_EGF) <- names_per
  names(all_grn_HGF) <- names_per
  names(all_new_grn) <- names_per
  names(all_ligand_target_matrix) <- names_per
  names(all_weighted_networks) <- names_per
}



# all_ligand_target_matrix <- list()
# all_new_grn <- c()
# for(per in c("0", "0.25", "0.5", "0.75", "1")){
#   #add LINCS GRN to default GRN NicheNet
#   new_gr_network <- gr_network %>% bind_rows(all_grn_OSM[per]) %>% bind_rows(all_grn_HGF[per]) %>% bind_rows(all_grn_EGF[per])
#   head(new_gr_network)
#   # aggregate the individual data sources in a weighted manner to obtain a weighted integrated signaling network
#   weighted_networks = construct_weighted_networks(lr_network = lr_network, sig_network = sig_network, gr_network = new_gr_network,
#                                                   source_weights_df = new_source_weights_df)
#   all_new_grn[per] <- as.data.frame(new_gr_network)
#   
#   #construct ligand_target_matrix
#   ligands <- list("EGF", "HGF", "OSM")
#   ligand_target_matrix <- construct_ligand_target_matrix(weighted_networks = weighted_networks, ligands = ligands)
#   all_ligand_target_matrix[[per]] <- ligand_target_matrix
# }

table(unique(gr_network_LINCS_OSM$to) %in% table1$DE_info$OSM$external_gene_name)
length(unique(table1$DE_info$HGF$external_gene_name))
```

## Evaluation NicheNet
# Transcriptional response prediction evaluation
```{r model evaluation with per, echo=TRUE, error=TRUE, warning=TRUE}
#one ligand (EGF, HGF, OSM)
expression_settings_validation_DE_one <- list(designs[[1]]$expression_settings_validation_DE$EGF,
                                designs[[1]]$expression_settings_validation_DE$HGF,
                                designs[[1]]$expression_settings_validation_DE$OSM)

names_one <- c("EGF", "HGF", "OSM")
names(expression_settings_validation_DE_one) <- names_one

all_performances <- c()
names_per <- c()
for(per in c("0", "0.25", "0.5", "0.75", "1")){
  ligand_target_matrix <- all_ligand_target_matrix[[per]]
  if(length(expression_settings_validation_DE_one) > 0){
    # Step 1: convert expression datasets to the required format to perform target gene prediction
    settings_one = expression_settings_validation_DE_one %>% lapply(convert_expression_settings_evaluation)
    settings_one = settings_one %>% discard(~length(.$from) > 1)
    designs[[1]]$settings_one <- settings_one
    #print(names(which(settings$response)))
    
    # Step 2: calculate the target gene prediction performances
    performances_one = settings_one %>% lapply(evaluate_target_prediction, ligand_target_matrix) %>% bind_rows() 
    designs[[1]]$performances_one <- as.data.frame(performances_one)
    
      
    # Visualize some classification evaluation metrics showing the target gene prediction performance
    performances_one = performances_one %>% dplyr::select(-aupr, -auc_iregulon,-pearson_log_pval,-spearman_log_pval , -sensitivity_roc, -specificity_roc) %>% gather(key = scorename, value = scorevalue, auroc:spearman)
    scorelabels = c(auroc = "AUROC", aupr_corrected = "AUPR (corrected)", auc_iregulon_corrected = "AUC-iRegulon (corrected)", 
                    pearson = "Pearson correlation", spearman = "Spearman's rank correlation", 
                    mean_rank_GST_log_pval = "Mean-rank gene-set enrichment")
    scorerandom = c(auroc = 0.5, aupr_corrected = 0, auc_iregulon_corrected = 0, pearson = 0, spearman = 0, mean_rank_GST_log_pval = 0) %>% data.frame(scorevalue = .) %>% rownames_to_column("scorename")
    names_per <- append(names_per, per)
    
    all_performances$per <- performances_one
    names(all_performances) <- names_per
      
    print(performances_one %>%
            mutate(model = "NicheNet") %>%
            ggplot() +
            geom_violin(aes(model, scorevalue, group = model, fill = model)) +
            geom_boxplot(aes(model, scorevalue, group = model), width = 0.05) +
            scale_y_continuous("Score target prediction") +
            facet_wrap(~scorename, scales = "free", labeller = as_labeller(scorelabels)) +
            geom_hline(aes(yintercept = scorevalue), data = scorerandom, linetype = 2, color = "red") +
            theme_bw() +
            ggtitle(paste("24h with ANANSE data only one ligand", per))
    )
  }
}

```

# ligand activity prediction evaluation

```{r ligand activity prediction evaluation with per, echo=TRUE, error=TRUE}
#one ligand
designs[[1]]$settings_one <- settings_one
names(settings_one) <- names_one

all_evaluation_ligand_predictions <- c()
names_per <- c()
for(per in c("0", "0.25", "0.5", "0.75", "1")){
  ligand_target_matrix <- all_ligand_target_matrix[[per]]
  if(length(settings_one) > 0){
    # convert expression datasets to correct format for ligand activity prediction
    all_ligands = settings_one %>% extract_ligands_from_settings(combination = FALSE) %>% unlist()
    settings_ligand_prediction_one = settings_one %>% convert_settings_ligand_prediction(all_ligands = all_ligands, validation = TRUE)
      
     # infer ligand importances: for all ligands of interest, we assess how well a ligand explains the differential expression in a specific dataset (and we do this for all datasets).
     ligand_importances_one = settings_ligand_prediction_one %>% lapply(get_single_ligand_importances, ligand_target_matrix) %>% bind_rows()
     ligand_importances_one <- do.call(data.frame, lapply(ligand_importances_one, function(value) replace(value, is.infinite(value), 100000)))
      
      # Look at predictive performance of single/individual importance measures to predict ligand activity: of all ligands tested, the ligand that is truly active in a dataset should get the highest activity score (i.e. best target gene prediction performance)
     evaluation_ligand_prediction_one = ligand_importances_one$setting %>% unique() %>% lapply(function(x){x}) %>% 
      lapply(wrapper_evaluate_single_importances_ligand_prediction, ligand_importances_one) %>%
      bind_rows() %>% inner_join(ligand_importances_one %>% distinct(setting, ligand))
     
      
    # Visualize some classification evaluation metrics showing the ligand activity prediction performance
    evaluation_ligand_prediction_one = evaluation_ligand_prediction_one %>% dplyr::select(-aupr, -sensitivity_roc, -specificity_roc,
                                                                                          -pearson, -spearman,
                                                                                          -mean_rank_GST_log_pval) %>% 
      gather(key = scorename, value = scorevalue, auroc:aupr_corrected)
    
    names_per <- append(names_per, per)
    all_evaluation_ligand_predictions$per <- evaluation_ligand_prediction_one
    names(all_performances) <- names_per
    
    scorelabels = c(auroc = "AUROC", aupr_corrected = "AUPR (corrected)")
    scorerandom = c(auroc = 0.5, aupr_corrected = 0) %>% data.frame(scorevalue = .) %>% rownames_to_column("scorename")
      
    print(evaluation_ligand_prediction_one %>%
       filter(importance_measure %in% c("auroc", "aupr_corrected", "mean_rank_GST_log_pval", "auc_iregulon_corrected", "pearson",
                                        "spearman")) %>%
      ggplot() +
      geom_violin(aes(importance_measure, scorevalue, group = importance_measure, fill = importance_measure)) +
      geom_boxplot(aes(importance_measure, scorevalue, group = importance_measure), width = 0.1) +
      scale_y_continuous("Evaluation ligand activity prediction") +
      scale_x_discrete("Ligand activity measure") +
      facet_wrap(~scorename, scales = "free", labeller=as_labeller(scorelabels)) +
      geom_hline(aes(yintercept=scorevalue), data=scorerandom, linetype = 2, color = "red") +
      theme_bw() +
      theme(axis.text.x = element_text(angle = 90))) +
      ggtitle(paste("24h with ANANSE data only one ligand", per))
  }
}

```

#Statsplots NicheNet against all ATAC data
```{r statsplots NicheNet, echo=TRUE, error=TRUE, fig.width=10, fig.height = 10}
performances_NicheNet["condition"] <- "NicheNet" 
performances_total["condition"] <- "ATAC" 
perf <- performances_NicheNet %>% rbind(performances_total)

au <- perf[which(perf$scorename == "auroc"),]
au_plot <- ggwithinstats(
  data = au,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC")) +
  ggtitle ("AUROC")

aupr <- perf[which(perf$scorename == "aupr_corrected"),]
aupr_plot <- ggwithinstats(
  data = aupr,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("AUPR")



gs <- perf[which(perf$scorename == "mean_rank_GST_log_pval"),]
gs_plot <- ggwithinstats(
  data = gs,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("mean rank GST log pval")

iregulon <- perf[which(perf$scorename == "auc_iregulon_corrected"),]
auc_plot <- ggwithinstats(
  data = iregulon,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("AUC")

sp <- perf[which(perf$scorename == "spearman"),]
sp_plot <- ggwithinstats(
  data = sp,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("Spearman")

pe <- perf[which(perf$scorename == "pearson"),]
pe_plot<- ggwithinstats(
  data = pe,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("Pearson")

print(plot_grid(au_plot, aupr_plot, gs_plot, auc_plot, pe_plot, sp_plot))
```

# Statsplots NicheNet 3 ligands against all ATAC data

```{r statsplots NicheNet ligands, echo=TRUE, error=TRUE, fig.width=10, fig.height = 10}
performances_NicheNet_ligands["condition"] <- "NicheNet" 
performances_total["condition"] <- "ATAC" 
perf <- performances_NicheNet_ligands %>% rbind(performances_total)

au <- perf[which(perf$scorename == "auroc"),]
au_plot <- ggwithinstats(
  data = au,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC")) +
  ggtitle ("AUROC")

aupr <- perf[which(perf$scorename == "aupr_corrected"),]
aupr_plot <- ggwithinstats(
  data = aupr,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("AUPR")



gs <- perf[which(perf$scorename == "mean_rank_GST_log_pval"),]
gs_plot <- ggwithinstats(
  data = gs,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("mean rank GST log pval")

iregulon <- perf[which(perf$scorename == "auc_iregulon_corrected"),]
auc_plot <- ggwithinstats(
  data = iregulon,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("AUC")

sp <- perf[which(perf$scorename == "spearman"),]
sp_plot <- ggwithinstats(
  data = sp,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("Spearman")

pe <- perf[which(perf$scorename == "pearson"),]
pe_plot<- ggwithinstats(
  data = pe,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("Pearson")

print(plot_grid(au_plot, aupr_plot, gs_plot, auc_plot, pe_plot, sp_plot))

```

#Statsplots with 0 to 100% DE genes against NicheNet with the 3 ligands

```{r statsplots 0, echo=TRUE, error=TRUE, fig.width=10, fig.height = 10}
performances_NicheNet_ligands["condition"] <- "NicheNet" 
all_performances$"0"["condition"] <- "ATAC" 
all_performances$"0.25"["condition"] <- "ATAC" 
all_performances$"0.5"["condition"] <- "ATAC" 
all_performances$"0.75"["condition"] <- "ATAC" 
all_performances$"1"["condition"] <- "ATAC" 

perf_0 <- performances_NicheNet_ligands %>% rbind(all_performances$"0")
perf_0.25 <- performances_NicheNet_ligands %>% rbind(all_performances$"0.25")
perf_0.5 <- performances_NicheNet_ligands %>% rbind(all_performances$"0.5")
perf_0.75 <- performances_NicheNet_ligands %>% rbind(all_performances$"0.75")
perf_1 <- performances_NicheNet_ligands %>% rbind(all_performances$"1")

au_0 <- perf_0[which(perf_0$scorename == "auroc"),]
au_0.25 <- perf_0.25[which(perf_0.25$scorename == "auroc"),]
au_0.5 <- perf_0.5[which(perf_0.5$scorename == "auroc"),]
au_0.75 <- perf_0.75[which(perf_0.75$scorename == "auroc"),]
au_1 <- perf_1[which(perf_1$scorename == "auroc"),]

aupr_0 <- perf_0[which(perf_0$scorename == "aupr_corrected"),]
aupr_0.25 <- perf_0.25[which(perf_0.25$scorename == "aupr_corrected"),]
aupr_0.5 <- perf_0.5[which(perf_0.5$scorename == "aupr_corrected"),]
aupr_0.75 <- perf_0.75[which(perf_0.75$scorename == "aupr_corrected"),]
aupr_1 <- perf_1[which(perf_1$scorename == "aupr_corrected"),]

gs_0 <- perf_0[which(perf_0$scorename == "mean_rank_GST_log_pval"),]
gs_0.25 <- perf_0.25[which(perf_0.25$scorename == "mean_rank_GST_log_pval"),]
gs_0.5 <- perf_0.5[which(perf_0.5$scorename == "mean_rank_GST_log_pval"),]
gs_0.75 <- perf_0.75[which(perf_0.75$scorename == "mean_rank_GST_log_pval"),]
gs_1 <- perf_1[which(perf_1$scorename == "mean_rank_GST_log_pval"),]

iregulon_0 <- perf_0[which(perf_0$scorename == "auc_iregulon_corrected"),]
iregulon_0.25 <- perf_0.25[which(perf_0.25$scorename == "auc_iregulon_corrected"),]
iregulon_0.5 <- perf_0.5[which(perf_0.5$scorename == "auc_iregulon_corrected"),]
iregulon_0.75 <- perf_0.75[which(perf_0.75$scorename == "auc_iregulon_corrected"),]
iregulon_1 <- perf_1[which(perf_1$scorename == "auc_iregulon_corrected"),]

sp_0 <- perf_0[which(perf_0$scorename == "spearman"),]
sp_0.25 <- perf_0.25[which(perf_0.25$scorename == "spearman"),]
sp_0.5 <- perf_0.5[which(perf_0.5$scorename == "spearman"),]
sp_0.75 <- perf_0.75[which(perf_0.75$scorename == "spearman"),]
sp_1 <- perf_1[which(perf_1$scorename == "spearman"),]


pe_0 <- perf_0[which(perf_0$scorename == "pearson"),]
pe_0.25 <- perf_0.25[which(perf_0.25$scorename == "pearson"),]
pe_0.5 <- perf_0.5[which(perf_0.5$scorename == "pearson"),]
pe_0.75<- perf_0.75[which(perf_0.75$scorename == "pearson"),]
pe_1 <- perf_1[which(perf_1$scorename == "pearson"),]

# Make AUC plots
au_plot_0 <- ggwithinstats(
  data = au_0,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC")) +
  ggtitle ("AUROC 0% DE genes")

au_plot_0.25 <- ggwithinstats(
  data = au_0.25,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC")) +
  ggtitle ("AUROC 25% DE genes")

au_plot_0.5 <- ggwithinstats(
  data = au_0.5,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC")) +
  ggtitle ("AUROC 50% DE genes")

au_plot_0.75 <- ggwithinstats(
  data = au_0.75,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC")) +
  ggtitle ("AUROC 75% DE genes")

au_plot_1 <- ggwithinstats(
  data = au_1,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC")) +
  ggtitle ("AUROC 100% DE genes")

print(plot_grid(au_plot_0, au_plot_0.25, au_plot_0.5, au_plot_0.75, au_plot_1))

#Make AUPR plots
aupr_plot_0 <- ggwithinstats(
  data = aupr_0,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("AUPR 0% DE genes")

aupr_plot_0.25 <- ggwithinstats(
  data = aupr_0.25,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("AUPR 25% DE genes")

aupr_plot_0.5 <- ggwithinstats(
  data = aupr_0.5,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("AUPR 50% DE genes")

aupr_plot_0.75 <- ggwithinstats(
  data = aupr_0.75,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("AUPR 75% DE genes")

aupr_plot_1 <- ggwithinstats(
  data = aupr_1,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("AUPR 100% DE genes")

print(plot_grid(aupr_plot_0, aupr_plot_0.25, aupr_plot_0.5, aupr_plot_0.75, aupr_plot_1))

#Make mean rank GST plots
gs_plot_0 <- ggwithinstats(
  data = gs_0,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("mean rank GST log pval 0% DE genes")

gs_plot_0.25 <- ggwithinstats(
  data = gs_0.25,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("mean rank GST log pval 25% DE genes")

gs_plot_0.5 <- ggwithinstats(
  data = gs_0.5,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("mean rank GST log pval 50% DE genes")

gs_plot_0.75 <- ggwithinstats(
  data = gs_0.75,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("mean rank GST log pval 75% DE genes")

gs_plot_1 <- ggwithinstats(
  data = gs_1,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("mean rank GST log pval 100% DE genes")

print(plot_grid(gs_plot_0, gs_plot_0.25, gs_plot_0.5, gs_plot_0.75, gs_plot_1))

#Make iregelon plots
iregulon_plot_0 <- ggwithinstats(
  data = iregulon_0,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("iregulon with 0% DE genes")

iregulon_plot_0.25 <- ggwithinstats(
  data = iregulon_0.25,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("iregulon with 25% DE genes")

iregulon_plot_0.5 <- ggwithinstats(
  data = iregulon_0.5,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("iregulon with 50% DE genes")

iregulon_plot_0.75 <- ggwithinstats(
  data = iregulon_0.75,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("iregulon with 75% DE genes")

iregulon_plot_1 <- ggwithinstats(
  data = iregulon_1,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("iregulon with 100% DE genes")

print(plot_grid(iregulon_plot_0, iregulon_plot_0.25, iregulon_plot_0.5, iregulon_plot_0.75, iregulon_plot_1))

#Make spearman plots
sp_plot_0 <- ggwithinstats(
  data = sp_0,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("Spearman with 0% DE genes")

sp_plot_0.25 <- ggwithinstats(
  data = sp_0.25,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("Spearman with 25% DE genes")

sp_plot_0.5 <- ggwithinstats(
  data = sp_0.5,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("Spearman with 50% DE genes")

sp_plot_0.75 <- ggwithinstats(
  data = sp_0.75,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("Spearman with 75% DE genes")

sp_plot_1 <- ggwithinstats(
  data = sp_1,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("Spearman with 100% DE genes")

print(plot_grid(sp_plot_0, sp_plot_0.25, sp_plot_0.5, sp_plot_0.75, sp_plot_1))

#Make Pearson plots
pe_plot_0 <- ggwithinstats(
  data = pe_0,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("Pearson with 0% DE genes")

pe_plot_0.25 <- ggwithinstats(
  data = pe_0.25,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("Pearson with 25% DE genes")

pe_plot_0.5 <- ggwithinstats(
  data = pe_0.5,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("Pearson with 50% DE genes")

pe_plot_0.75 <- ggwithinstats(
  data = pe_0.75,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("Pearson with 75% DE genes")

pe_plot_1 <- ggwithinstats(
  data = pe_1,
  x    = condition,
  y    = scorevalue,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("Pearson with 100% DE genes")

print(plot_grid(pe_plot_0, pe_plot_0.25, pe_plot_0.5, pe_plot_0.75, pe_plot_1))
```


```{r differences ligand_target_matrix}
# eq <- ligand_target_matrix_NicheNet_ligands[,"EGF"]==ligand_target_matrix_total["EGF",]# avoid to later compute this twice
# eee <- as.matrix(ligand_target_matrix_NicheNet_ligands[,"EGF"])
# ee_total <- as.matrix(ligand_target_matrix_total[,"EGF"])
# 
# equals <- eee[which(eee==ee_total),]
# which(equals != 0) #only 0s found that are the same
# ifelse(eq, 0, mat2) # get the desired matrix
# round(sum(!eq)/length(eq)*100, 2) # get the percentage of non equal values
# #[1] 100%, meaning all the targets are different
# 
# k <- as.matrix(all_ligand_target_matrix$"0"[,"EGF"])
# same <- as.matrix(k[which(all_ligand_target_matrix$"0"[,"EGF"]==all_ligand_target_matrix$"1"[,"EGF"]),]) #all the same
# generics::intersect(ligand_target_matrix_NicheNet_ligands, ligand_target_matrix_total)
# generics::intersect(all_ligand_target_matrix$"0", all_ligand_target_matrix$"0.25")
# 
# 
# sum((!which(ligand_target_matrix_NicheNet_ligands == ligand_target_matrix_total))=="FALSE")
# sum((!which(all_ligand_target_matrix$"0.25" == all_ligand_target_matrix$"0.5"))=="FALSE")
# sum((!which(all_ligand_target_matrix$"0.25" == all_ligand_target_matrix$"0.75"))=="FALSE")
# sum((!which(all_ligand_target_matrix$"0.25" == all_ligand_target_matrix$"1"))=="FALSE")
# sum((!which(all_ligand_target_matrix$"0.25" == all_ligand_target_matrix$"0"))=="FALSE")
# sum((!which(all_ligand_target_matrix$"0.25" == all_ligand_target_matrix$"0.5"))=="FALSE")
# sum((!which(all_ligand_target_matrix$"0.25" == all_ligand_target_matrix$"0.75"))=="FALSE")
# sum((!which(all_ligand_target_matrix$"0.25" == all_ligand_target_matrix$"1"))=="FALSE")
# sum((!which(all_ligand_target_matrix$"0.5" == all_ligand_target_matrix$"1"))=="FALSE")
# sum((!which(all_ligand_target_matrix$"0.5" == all_ligand_target_matrix$"0.75"))=="FALSE")

```
```{r check for differences in weighted networks or ligand target matrices}
# sum((all_weighted_networks$"0"$gr$weight == all_weighted_networks$"0.25"$gr$weight)=="FALSE")
# sum((all_ligand_target_matrix$"0" == all_ligand_target_matrix$"0.25")=="FALSE")
# sum((performances_NicheNet_ligands == performances_total) == "FALSE")
```

#Evaluation ligand_target_prediction

```{r statsplots evaluation try, echo=TRUE, error=TRUE, fig.width=10, fig.height = 10}
evaluation_ligand_prediction_NicheNet_ligands["condition"] <- "NicheNet_ligands" 
evaluation_ligand_prediction_total["condition"] <- "ATAC" 
eva <- evaluation_ligand_prediction_NicheNet_ligands %>% rbind(evaluation_ligand_prediction_total)

sub_auroc <- eva[which(eva$scorename == "auroc"),]
sub_aupr <- eva[which(eva$scorename == "aupr_corrected"),]

au <- sub_auroc[which(sub_auroc$importance_measure == "auroc"),]
au_plot <- ggwithinstats(
  data = au,
  x    = condition,
  y    = pearson_log_pval,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet_ligands", "ATAC")) +
  ggtitle ("AUROC")

aupr <-  sub_auroc[which(sub_auroc$importance_measure == "aupr_corrected"),]
aupr_plot <- ggwithinstats(
  data = aupr,
  x    = condition,
  y    = pearson_log_pval,
  type = "np"
) +
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("AUPR")



gs <-  sub_auroc[which(sub_auroc$importance_measure == "mean_rank_GST_log_pval"),]
gs_plot <- ggwithinstats(
  data = gs,
  x    = condition,
  y    = pearson_log_pval,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("mean rank GST log pval")

iregulon <-  sub_auroc[which(sub_auroc$importance_measure == "auc_iregulon_corrected"),]
auc_plot <- ggwithinstats(
  data = iregulon,
  x    = condition,
  y    = pearson_log_pval,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("AUC")

sp <-  sub_auroc[which(sub_auroc$importance_measure == "spearman"),]
sp_plot <- ggwithinstats(
  data = sp,
  x    = condition,
  y    = pearson_log_pval,
  type = "np"
)+
  scale_x_discrete(limits = c("NicheNet", "ATAC"))+
  ggtitle ("Spearman")

pe <-  sub_auroc[which(sub_auroc$importance_measure == "pearson"),]
pe_plot<- ggwithinstats(
  data = pe,
  x    = condition,
  y    = pearson_log_pval,
  type = "np"
)+
  scale_x_discrete(limits = c("total", "ATAC"))+
  ggtitle ("Pearson")

print(plot_grid(au_plot, aupr_plot, gs_plot, auc_plot, pe_plot, sp_plot))



 # print(evaluation_ligand_prediction_NicheNet_ligands %>%
 #     filter(importance_measure %in% c("auroc", "aupr_corrected", "mean_rank_GST_log_pval", "auc_iregulon_corrected", "pearson", "spearman")) %>%
 #    ggplot() +
 #    geom_violin(aes(importance_measure, scorevalue, group = importance_measure, fill = importance_measure)) +
 #    geom_boxplot(aes(importance_measure, scorevalue, group = importance_measure), width = 0.1) +
 #    scale_y_continuous("Evaluation ligand activity prediction") +
 #    scale_x_discrete("Ligand activity measure") +
 #    facet_wrap(~scorename, scales = "free", labeller = as_labeller(scorelabels)) )
 #    # geom_hline(aes(yintercept = scorevalue), data = scorerandom, linetype = 2, color = "red") +
 #    # theme_bw() +
 #    # theme(axis.text.x = element_text(angle = 90))) +
 #    # ggtitle("24h with ANANSE data only one ligand")

```
